// @atproto/oauth-provider/dist/oauth-verifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuthVerifier = exports.Keyset = exports.DpopNonce = exports.AccessTokenType = void 0;
const jwk_1 = require("@atproto/jwk");
Object.defineProperty(exports, "Keyset", { enumerable: true, get: function () { return jwk_1.Keyset; } });
const oauth_types_1 = require("@atproto/oauth-types");
const access_token_type_js_1 = require("./access-token/access-token-type.js");
Object.defineProperty(exports, "AccessTokenType", { enumerable: true, get: function () { return access_token_type_js_1.AccessTokenType; } });
const dpop_manager_js_1 = require("./dpop/dpop-manager.js");
const dpop_nonce_js_1 = require("./dpop/dpop-nonce.js");
Object.defineProperty(exports, "DpopNonce", { enumerable: true, get: function () { return dpop_nonce_js_1.DpopNonce; } });
const invalid_dpop_proof_error_js_1 = require("./errors/invalid-dpop-proof-error.js");
const invalid_token_error_js_1 = require("./errors/invalid-token-error.js");
const use_dpop_nonce_error_js_1 = require("./errors/use-dpop-nonce-error.js");
const www_authenticate_error_js_1 = require("./errors/www-authenticate-error.js");
const authorization_header_js_1 = require("./lib/util/authorization-header.js");
const replay_manager_js_1 = require("./replay/replay-manager.js");
const replay_store_memory_js_1 = require("./replay/replay-store-memory.js");
const replay_store_redis_js_1 = require("./replay/replay-store-redis.js");
const signer_js_1 = require("./signer/signer.js");
const verify_token_claims_js_1 = require("./token/verify-token-claims.js");
class OAuthVerifier {
    issuer;
    keyset;
    accessTokenType;
    dpopManager;
    replayManager;
    signer;
    constructor({ redis, issuer, keyset, replayStore = redis != null
        ? new replay_store_redis_js_1.ReplayStoreRedis({ redis })
        : new replay_store_memory_js_1.ReplayStoreMemory(), accessTokenType = access_token_type_js_1.AccessTokenType.jwt, ...rest }) {
        const dpopMgrOptions = rest;
        const issuerParsed = oauth_types_1.oauthIssuerIdentifierSchema.parse(issuer);
        const issuerUrl = new URL(issuerParsed);
        // TODO (?) support issuer with path
        if (issuerUrl.pathname !== '/') {
            throw new TypeError(`"issuer" must be an URL with no path, search or hash (${issuerUrl})`);
        }
        this.issuer = issuerParsed;
        this.keyset = keyset instanceof jwk_1.Keyset ? keyset : new jwk_1.Keyset(keyset);
        this.accessTokenType = accessTokenType;
        this.dpopManager = new dpop_manager_js_1.DpopManager(dpopMgrOptions);
        this.replayManager = new replay_manager_js_1.ReplayManager(replayStore);
        this.signer = new signer_js_1.Signer(this.issuer, this.keyset);
    }
    nextDpopNonce() {
        return this.dpopManager.nextNonce();
    }
    async checkDpopProof(proof, htm, htu, accessToken) {
        if (proof === undefined)
            return null;
        const { payload, jkt } = await this.dpopManager.checkProof(proof, htm, htu, accessToken);
        const unique = await this.replayManager.uniqueDpop(payload.jti);
        if (!unique)
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP proof jti is not unique');
        return jkt;
    }
    assertTokenTypeAllowed(tokenType, accessTokenType) {
        if (this.accessTokenType !== access_token_type_js_1.AccessTokenType.auto &&
            this.accessTokenType !== accessTokenType) {
            throw new invalid_token_error_js_1.InvalidTokenError(tokenType, `Invalid token type`);
        }
    }
    async authenticateToken(tokenType, token, dpopJkt, verifyOptions) {
        if (!(0, jwk_1.isSignedJwt)(token)) {
            throw new invalid_token_error_js_1.InvalidTokenError(tokenType, `Malformed token`);
        }
        this.assertTokenTypeAllowed(tokenType, access_token_type_js_1.AccessTokenType.jwt);
        const { payload } = await this.signer
            .verifyAccessToken(token)
            .catch((err) => {
            throw invalid_token_error_js_1.InvalidTokenError.from(err, tokenType);
        });
        return (0, verify_token_claims_js_1.verifyTokenClaims)(token, payload.jti, tokenType, dpopJkt, payload, verifyOptions);
    }
    async authenticateRequest(method, url, headers, verifyOptions) {
        const [tokenType, token] = (0, authorization_header_js_1.parseAuthorizationHeader)(headers.authorization);
        try {
            const dpopJkt = await this.checkDpopProof(headers.dpop, method, url, token);
            if (tokenType === 'DPoP' && !dpopJkt) {
                throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError(`DPoP proof required`);
            }
            return await this.authenticateToken(tokenType, token, dpopJkt, verifyOptions);
        }
        catch (err) {
            if (err instanceof use_dpop_nonce_error_js_1.UseDpopNonceError)
                throw err.toWwwAuthenticateError();
            if (err instanceof www_authenticate_error_js_1.WWWAuthenticateError)
                throw err;
            throw invalid_token_error_js_1.InvalidTokenError.from(err, tokenType);
        }
    }
}
exports.OAuthVerifier = OAuthVerifier;
