// @atproto/oauth-provider/dist/output/build-customization-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customizationSchema = exports.brandingConfigSchema = exports.linkDefinitionSchema = exports.linkRelSchema = exports.localizedStringSchema = exports.colorsDefinitionSchema = exports.colorNameSchema = exports.colorNames = exports.hcaptchaConfigSchema = void 0;
exports.buildCustomizationData = buildCustomizationData;
exports.buildCustomizationCss = buildCustomizationCss;
const zod_1 = require("zod");
const hcaptcha_js_1 = require("../lib/hcaptcha.js");
const build_document_js_1 = require("../lib/html/build-document.js");
const locale_js_1 = require("../lib/locale.js");
var hcaptcha_js_2 = require("../lib/hcaptcha.js");
Object.defineProperty(exports, "hcaptchaConfigSchema", { enumerable: true, get: function () { return hcaptcha_js_2.hcaptchaConfigSchema; } });
// Matches colors defined in tailwind.config.js
exports.colorNames = ['brand', 'error', 'warning', 'success'];
exports.colorNameSchema = zod_1.z.enum(exports.colorNames);
const parsedColorSchema = zod_1.z.string().transform((value, ctx) => {
    try {
        const { r, g, b, a } = parseColor(value);
        if (a != null) {
            ctx.addIssue({
                code: zod_1.z.ZodIssueCode.custom,
                message: 'Alpha values are not supported',
            });
        }
        return { r, g, b };
    }
    catch (e) {
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: e instanceof Error ? e.message : 'Invalid color value',
        });
        // Won't actually be used (since an issue was added):
        return { r: 0, g: 0, b: 0 };
    }
});
exports.colorsDefinitionSchema = zod_1.z.record(exports.colorNameSchema, parsedColorSchema.optional());
exports.localizedStringSchema = zod_1.z.union([
    zod_1.z.string(),
    locale_js_1.multiLangStringSchema,
]);
exports.linkRelSchema = zod_1.z.string().refine(build_document_js_1.isLinkRel, 'Invalid link rel');
exports.linkDefinitionSchema = zod_1.z.object({
    title: exports.localizedStringSchema,
    href: zod_1.z.string().url(),
    rel: exports.linkRelSchema.optional(),
});
/**
 * Aesthetic customization
 */
exports.brandingConfigSchema = zod_1.z.object({
    name: zod_1.z.string().optional(),
    logo: zod_1.z.string().optional(),
    colors: exports.colorsDefinitionSchema.optional(),
    links: zod_1.z.array(exports.linkDefinitionSchema).optional(),
});
exports.customizationSchema = zod_1.z.object({
    /**
     * Available user domains that can be used to sign up. A non-empty array
     * is required to enable the sign-up feature.
     */
    availableUserDomains: zod_1.z.array(zod_1.z.string()).optional(),
    /**
     * UI customizations
     */
    branding: exports.brandingConfigSchema.optional(),
    /**
     * Is an invite code required to sign up?
     */
    inviteCodeRequired: zod_1.z.boolean().optional(),
    /**
     * Enables hCaptcha during sign-up.
     */
    hcaptcha: hcaptcha_js_1.hcaptchaConfigSchema.optional(),
});
function buildCustomizationData({ branding, availableUserDomains, inviteCodeRequired, hcaptcha, }) {
    // @NOTE the front end does not need colors here as they will be injected as
    // CSS variables.
    // @NOTE We only copy the values explicitly needed to avoid leaking sensitive
    // data (in case the caller passed more than what we expect).
    return {
        availableUserDomains,
        inviteCodeRequired,
        hcaptchaSiteKey: hcaptcha?.siteKey,
        name: branding?.name,
        logo: branding?.logo,
        links: branding?.links,
    };
}
function buildCustomizationCss({ branding }) {
    const vars = Array.from(buildCustomizationVars(branding));
    if (vars.length)
        return `:root { ${vars.join(' ')} }`;
    return '';
}
function* buildCustomizationVars(branding) {
    if (branding?.colors) {
        for (const name of exports.colorNames) {
            const value = branding.colors[name];
            if (!value)
                continue; // Skip missing colors
            const { r, g, b } = value;
            const contrast = computeLuma({ r, g, b }) > 128 ? '0 0 0' : '255 255 255';
            yield `--color-${name}: ${r} ${g} ${b};`;
            yield `--color-${name}-c: ${contrast};`;
        }
    }
}
function parseColor(color) {
    if (color.startsWith('#')) {
        return parseHexColor(color);
    }
    if (color.startsWith('rgba(')) {
        return parseRgbaColor(color);
    }
    if (color.startsWith('rgb(')) {
        return parseRgbColor(color);
    }
    // Should never happen (as long as the input is a validated WebColor)
    throw new TypeError(`Invalid color value: ${color}`);
}
function parseHexColor(v) {
    // parseInt('az', 16) does not return NaN so we need to check the format
    if (!/^#[0-9a-f]+$/i.test(v)) {
        throw new TypeError(`Invalid hex color value: ${v}`);
    }
    if (v.length === 4 || v.length === 5) {
        const r = parseUi8Hex(v.slice(1, 2));
        const g = parseUi8Hex(v.slice(2, 3));
        const b = parseUi8Hex(v.slice(3, 4));
        const a = v.length > 4 ? parseUi8Hex(v.slice(4, 5)) : undefined;
        return { r, g, b, a };
    }
    if (v.length === 7 || v.length === 9) {
        const r = parseUi8Hex(v.slice(1, 3));
        const g = parseUi8Hex(v.slice(3, 5));
        const b = parseUi8Hex(v.slice(5, 7));
        const a = v.length > 8 ? parseUi8Hex(v.slice(7, 9)) : undefined;
        return { r, g, b, a };
    }
    throw new TypeError(`Invalid hex color value: ${v}`);
}
function parseRgbColor(v) {
    const matches = v.match(/^\s*rgb\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*$/);
    if (!matches)
        throw new TypeError(`Invalid rgb color value: ${v}`);
    const r = parseUi8Dec(matches[1]);
    const g = parseUi8Dec(matches[2]);
    const b = parseUi8Dec(matches[3]);
    return { r, g, b };
}
function parseRgbaColor(v) {
    const matches = v.match(/^\s*rgba\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*$/);
    if (!matches)
        throw new TypeError(`Invalid rgba color value: ${v}`);
    const r = parseUi8Dec(matches[1]);
    const g = parseUi8Dec(matches[2]);
    const b = parseUi8Dec(matches[3]);
    const a = parseUi8Dec(matches[4]);
    return { r, g, b, a };
}
function computeLuma({ r, g, b }) {
    return 0.299 * r + 0.587 * g + 0.114 * b;
}
function parseUi8Hex(v) {
    return asUi8(parseInt(v, 16));
}
function parseUi8Dec(v) {
    return asUi8(parseInt(v, 10));
}
function asUi8(v) {
    if (v >= 0 && v <= 255 && v === (v | 0))
        return v;
    throw new TypeError(`Invalid color component "${v}" (expected an integer between 0 and 255)`);
}
