// @atproto/oauth-provider/dist/output/build-error-payload.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildErrorStatus = buildErrorStatus;
exports.buildErrorPayload = buildErrorPayload;
const jose_1 = require("jose");
const zod_1 = require("zod");
const jwk_1 = require("@atproto/jwk");
const oauth_error_js_1 = require("../errors/oauth-error.js");
const { JOSEError } = jose_1.errors;
const INVALID_REQUEST = 'invalid_request';
const SERVER_ERROR = 'server_error';
function buildErrorStatus(error) {
    if (error instanceof oauth_error_js_1.OAuthError) {
        return error.statusCode;
    }
    if (error instanceof jwk_1.JwtVerifyError) {
        return 400;
    }
    if (error instanceof zod_1.ZodError) {
        return 400;
    }
    if (error instanceof JOSEError) {
        return 400;
    }
    if (error instanceof TypeError) {
        return 400;
    }
    if (isBoom(error)) {
        return error.output.statusCode;
    }
    if (isXrpcError(error)) {
        return error.type;
    }
    const status = error?.status;
    if (typeof status === 'number' &&
        status === (status | 0) &&
        status >= 400 &&
        status < 600) {
        return status;
    }
    return 500;
}
function buildErrorPayload(error) {
    if (error instanceof oauth_error_js_1.OAuthError) {
        return error.toJSON();
    }
    if (error instanceof zod_1.ZodError) {
        return {
            error: INVALID_REQUEST,
            error_description: error.issues[0]?.message || 'Invalid request',
        };
    }
    if (error instanceof JOSEError) {
        return {
            error: INVALID_REQUEST,
            error_description: error.message,
        };
    }
    if (error instanceof TypeError) {
        return {
            error: INVALID_REQUEST,
            error_description: error.message,
        };
    }
    if (isBoom(error)) {
        return {
            error: error.output.statusCode <= 500 ? INVALID_REQUEST : SERVER_ERROR,
            error_description: error.output.statusCode <= 500
                ? isPayloadLike(error.output?.payload)
                    ? error.output.payload.message
                    : error.message
                : 'Server error',
        };
    }
    if (isXrpcError(error)) {
        return {
            error: error.type <= 500 ? INVALID_REQUEST : SERVER_ERROR,
            error_description: error.payload.message,
        };
    }
    const status = buildErrorStatus(error);
    return {
        error: status < 500 ? INVALID_REQUEST : SERVER_ERROR,
        error_description: error instanceof Error && error?.expose === true
            ? error.message
            : 'Server error',
    };
}
function isBoom(v) {
    return (v instanceof Error &&
        v.isBoom === true &&
        isHttpErrorCode(v['output']?.['statusCode']));
}
function isXrpcError(v) {
    return (v instanceof Error &&
        isHttpErrorCode(v['type']) &&
        isPayloadLike(v['payload']));
}
function isHttpErrorCode(v) {
    return typeof v === 'number' && v >= 400 && v < 600 && v === (v | 0);
}
function isPayloadLike(v) {
    return (v != null &&
        typeof v === 'object' &&
        typeof v['error'] === 'string' &&
        typeof v['message'] === 'string');
}
