// @atproto/oauth-provider/dist/output/output-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutputManager = void 0;
const oauth_provider_ui_1 = require("@atproto/oauth-provider-ui");
const assets_middleware_js_1 = require("../assets/assets-middleware.js");
const index_js_1 = require("../lib/csp/index.js");
const index_js_2 = require("../lib/html/index.js");
const security_headers_js_1 = require("../lib/http/security-headers.js");
const locale_js_1 = require("../lib/locale.js");
const backend_data_js_1 = require("./backend-data.js");
const build_authorize_data_js_1 = require("./build-authorize-data.js");
const build_customization_data_js_1 = require("./build-customization-data.js");
const build_error_data_js_1 = require("./build-error-data.js");
const build_error_payload_js_1 = require("./build-error-payload.js");
const send_web_page_js_1 = require("./send-web-page.js");
const BASE_CSP = {
    // API calls are made to the same origin
    'connect-src': ["'self'"],
    // Allow loading of PDS logo & User avatars
    'img-src': ['data:', 'https:'],
    // Prevent embedding in iframes
    'frame-ancestors': ["'none'"],
};
/**
 * @see {@link https://docs.hcaptcha.com/#content-security-policy-settings}
 */
const HCAPTCHA_CSP = {
    'script-src': ['https://hcaptcha.com', 'https://*.hcaptcha.com'],
    'frame-src': ['https://hcaptcha.com', 'https://*.hcaptcha.com'],
    'style-src': ['https://hcaptcha.com', 'https://*.hcaptcha.com'],
    'connect-src': ['https://hcaptcha.com', 'https://*.hcaptcha.com'],
};
class OutputManager {
    links;
    meta = [
        { name: 'robots', content: 'noindex' },
        { name: 'description', content: 'ATProto OAuth authorization page' },
    ];
    csp;
    coep;
    customizationData;
    customizationCss;
    constructor(customization) {
        this.links = customization.branding?.links;
        // "cache" these:
        this.customizationData = (0, build_customization_data_js_1.buildCustomizationData)(customization);
        this.customizationCss = (0, index_js_2.cssCode)((0, build_customization_data_js_1.buildCustomizationCss)(customization));
        this.csp = (0, index_js_1.mergeCsp)(BASE_CSP, customization?.hcaptcha ? HCAPTCHA_CSP : undefined);
        // Because we are loading avatar images from external sources, that might
        // not have CORP headers set, we need to use at least "credentialless".
        this.coep = customization?.hcaptcha
            ? // https://github.com/hCaptcha/react-hcaptcha/issues/259
                // @TODO Remove the use of `unsafeNone` once the issue above is resolved
                security_headers_js_1.CrossOriginEmbedderPolicy.unsafeNone
            : security_headers_js_1.CrossOriginEmbedderPolicy.credentialless;
    }
    buildAssets(name, backendData) {
        return {
            scripts: [
                (0, backend_data_js_1.declareBackendData)(backendData),
                // After backend injected data
                ...Array.from(oauth_provider_ui_1.assets)
                    .filter(([, item]) => item.type === 'chunk' && item.isEntry && item.name === name)
                    .map(([filename]) => ({ url: (0, assets_middleware_js_1.buildAssetUrl)(filename) })),
            ],
            styles: [
                ...Array.from(oauth_provider_ui_1.assets)
                    .filter(([, item]) => item.mime === 'text/css')
                    .map(([filename]) => ({ url: (0, assets_middleware_js_1.buildAssetUrl)(filename) })),
                // Last (to be able to override the default styles)
                this.customizationCss,
            ],
        };
    }
    async sendAuthorizePage(res, data, options) {
        const locale = negotiateLocale(data.parameters.ui_locales?.split(' ') ?? options?.preferredLocales);
        const { scripts, styles } = this.buildAssets('authorization-page', {
            __availableLocales: locale_js_1.AVAILABLE_LOCALES,
            __customizationData: this.customizationData,
            __authorizeData: (0, build_authorize_data_js_1.buildAuthorizeData)(data),
        });
        return (0, send_web_page_js_1.sendWebPage)(res, {
            scripts,
            styles,
            meta: this.meta,
            links: this.buildLinks(locale),
            htmlAttrs: { lang: locale },
            body: (0, index_js_2.html) `<div id="root"></div>`,
            csp: this.csp,
            coep: this.coep,
        });
    }
    async sendErrorPage(res, err, options) {
        const locale = negotiateLocale(options?.preferredLocales);
        const { scripts, styles } = this.buildAssets('error-page', {
            __availableLocales: locale_js_1.AVAILABLE_LOCALES,
            __customizationData: this.customizationData,
            __errorData: (0, build_error_data_js_1.buildErrorData)(err),
        });
        return (0, send_web_page_js_1.sendWebPage)(res, {
            status: (0, build_error_payload_js_1.buildErrorStatus)(err),
            scripts,
            styles,
            meta: this.meta,
            links: this.buildLinks(locale),
            htmlAttrs: { lang: locale },
            body: (0, index_js_2.html) `<div id="root"></div>`,
            csp: this.csp,
            coep: this.coep,
        });
    }
    buildLinks(locale) {
        return this.links
            ?.map(({ rel, href, title }) => (0, index_js_2.isLinkRel)(rel)
            ? typeof title === 'string'
                ? { href, rel, title }
                : { href, rel, title: title[locale] || title.en }
            : undefined)
            .filter((v) => v != null);
    }
}
exports.OutputManager = OutputManager;
function negotiateLocale(desiredLocales) {
    if (desiredLocales) {
        for (const locale of desiredLocales) {
            if (locale === '*')
                break; // use default
            if ((0, locale_js_1.isAvailableLocale)(locale))
                return locale;
        }
    }
    return 'en';
}
