// @atproto/oauth-provider/dist/output/send-authorize-redirect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendAuthorizeRedirect = sendAuthorizeRedirect;
const invalid_request_error_js_1 = require("../errors/invalid-request-error.js");
const index_js_1 = require("../lib/html/index.js");
const send_web_page_js_1 = require("./send-web-page.js");
// https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-11#section-7.5.4
const REDIRECT_STATUS_CODE = 303;
const SUCCESS_REDIRECT_KEYS = [
    'code',
    'id_token',
    'access_token',
    'expires_in',
    'token_type',
];
const ERROR_REDIRECT_KEYS = ['error', 'error_description', 'error_uri'];
async function sendAuthorizeRedirect(res, result) {
    const { issuer, parameters, redirect } = result;
    const uri = parameters.redirect_uri;
    if (!uri)
        throw new invalid_request_error_js_1.InvalidRequestError('No redirect_uri');
    const mode = parameters.response_mode || 'query'; // @TODO: default should depend on response_type
    const entries = [
        ['iss', issuer], // rfc9207
    ];
    if (parameters.state != null) {
        entries.push(['state', parameters.state]);
    }
    const keys = 'code' in redirect ? SUCCESS_REDIRECT_KEYS : ERROR_REDIRECT_KEYS;
    for (const key of keys) {
        const value = redirect[key];
        if (value != null)
            entries.push([key, value]);
    }
    res.setHeader('Cache-Control', 'no-store');
    switch (mode) {
        case 'query':
            return writeQuery(res, uri, entries);
        case 'fragment':
            return writeFragment(res, uri, entries);
        case 'form_post':
            return writeFormPost(res, uri, entries);
    }
    // @ts-expect-error fool proof
    throw new Error(`Unsupported mode: ${mode}`);
}
function writeQuery(res, uri, entries) {
    const url = new URL(uri);
    for (const [key, value] of entries)
        url.searchParams.set(key, value);
    res.writeHead(REDIRECT_STATUS_CODE, { Location: url.href }).end();
}
function writeFragment(res, uri, entries) {
    const url = new URL(uri);
    const searchParams = new URLSearchParams();
    for (const [key, value] of entries)
        searchParams.set(key, value);
    url.hash = searchParams.toString();
    res.writeHead(REDIRECT_STATUS_CODE, { Location: url.href }).end();
}
async function writeFormPost(res, uri, entries) {
    // Prevent the Chrome from caching this page
    // see: https://latesthackingnews.com/2023/12/12/google-updates-chrome-bfcache-for-faster-page-viewing/
    res.setHeader('Set-Cookie', `bfCacheBypass=foo; max-age=1; SameSite=Lax`);
    res.setHeader('Cache-Control', 'no-store');
    res.setHeader('Permissions-Policy', 'otp-credentials=*, document-domain=()');
    return (0, send_web_page_js_1.sendWebPage)(res, {
        htmlAttrs: { lang: 'en' },
        body: (0, index_js_1.html) `
      <form method="post" action="${uri}">
        ${entries.map(([key, value]) => [
            (0, index_js_1.html) `<input type="hidden" name="${key}" value="${value}" />`,
        ])}
        <input type="submit" value="Continue" />
      </form>
    `,
        scripts: [(0, index_js_1.js) `document.forms[0].submit();`],
    });
}
