// @atproto/oauth-provider/dist/signer/signer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Signer = void 0;
const date_js_1 = require("../lib/util/date.js");
const signed_token_payload_js_1 = require("./signed-token-payload.js");
class Signer {
    issuer;
    keyset;
    constructor(issuer, keyset) {
        this.issuer = issuer;
        this.keyset = keyset;
    }
    async verify(token, options) {
        return this.keyset.verifyJwt(token, {
            ...options,
            issuer: [this.issuer],
        });
    }
    async sign(signHeader, payload) {
        return this.keyset.createJwt(signHeader, async (protectedHeader, key) => ({
            ...(typeof payload === 'function'
                ? await payload(protectedHeader, key)
                : payload),
            iss: this.issuer,
        }));
    }
    async accessToken(client, parameters, options) {
        return this.sign({
            // https://datatracker.ietf.org/doc/html/rfc9068#section-2.1
            alg: options.alg,
            typ: 'at+jwt',
        }, {
            aud: options.aud,
            iat: (0, date_js_1.dateToEpoch)(options?.iat),
            exp: (0, date_js_1.dateToEpoch)(options.exp),
            sub: options.sub,
            jti: options.jti,
            cnf: options.cnf,
            // https://datatracker.ietf.org/doc/html/rfc8693#section-4.3
            client_id: client.id,
            scope: parameters.scope,
            authorization_details: options.authorization_details,
        });
    }
    async verifyAccessToken(token, options) {
        const result = await this.verify(token, { ...options, typ: 'at+jwt' });
        return {
            protectedHeader: result.protectedHeader,
            payload: signed_token_payload_js_1.signedTokenPayloadSchema.parse(result.payload),
        };
    }
}
exports.Signer = Signer;
