// @atproto/oauth-provider/dist/client/client-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientManager = void 0;
const jwk_1 = require("@atproto/jwk");
const oauth_types_1 = require("@atproto/oauth-types");
const fetch_1 = require("@atproto-labs/fetch");
const pipe_1 = require("@atproto-labs/pipe");
const simple_store_1 = require("@atproto-labs/simple-store");
const invalid_client_metadata_error_js_1 = require("../errors/invalid-client-metadata-error.js");
const invalid_redirect_uri_error_js_1 = require("../errors/invalid-redirect-uri-error.js");
const function_js_1 = require("../lib/util/function.js");
const hostname_js_1 = require("../lib/util/hostname.js");
const client_utils_js_1 = require("./client-utils.js");
const client_js_1 = require("./client.js");
const fetchMetadataHandler = (0, pipe_1.pipe)((0, fetch_1.fetchOkProcessor)(), 
// https://drafts.aaronpk.com/draft-parecki-oauth-client-id-metadata-document/draft-parecki-oauth-client-id-metadata-document.html#section-4.1
(0, fetch_1.fetchJsonProcessor)('application/json', true), (0, fetch_1.fetchJsonZodProcessor)(oauth_types_1.oauthClientMetadataSchema));
const fetchJwksHandler = (0, pipe_1.pipe)((0, fetch_1.fetchOkProcessor)(), (0, fetch_1.fetchJsonProcessor)('application/json', false), (0, fetch_1.fetchJsonZodProcessor)(jwk_1.jwksSchema));
class ClientManager {
    serverMetadata;
    keyset;
    hooks;
    store;
    loopbackMetadata;
    jwks;
    metadataGetter;
    constructor(serverMetadata, keyset, hooks, store, loopbackMetadata = null, safeFetch, clientJwksCache, clientMetadataCache) {
        this.serverMetadata = serverMetadata;
        this.keyset = keyset;
        this.hooks = hooks;
        this.store = store;
        this.loopbackMetadata = loopbackMetadata;
        const fetch = (0, fetch_1.bindFetch)(safeFetch);
        this.jwks = new simple_store_1.CachedGetter(async (uri, options) => {
            const jwks = await fetch(buildJsonGetRequest(uri, options)).then(fetchJwksHandler);
            return jwks;
        }, clientJwksCache);
        this.metadataGetter = new simple_store_1.CachedGetter(async (uri, options) => {
            const metadata = await fetch(buildJsonGetRequest(uri, options)).then(fetchMetadataHandler);
            // Validate within the getter to avoid caching invalid metadata
            return this.validateClientMetadata(uri, metadata);
        }, clientMetadataCache);
    }
    /**
     *
     * @see {@link https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2 OIDC Client Registration}
     */
    async getClient(clientId) {
        const metadata = await this.getClientMetadata(clientId).catch((err) => {
            throw invalid_client_metadata_error_js_1.InvalidClientMetadataError.from(err, `Unable to obtain client metadata for "${clientId}"`);
        });
        const jwks = metadata.jwks_uri
            ? await this.jwks.get(metadata.jwks_uri).catch((err) => {
                throw invalid_client_metadata_error_js_1.InvalidClientMetadataError.from(err, `Unable to obtain jwks from "${metadata.jwks_uri}" for "${clientId}"`);
            })
            : undefined;
        const partialInfo = await (0, function_js_1.callAsync)(this.hooks.getClientInfo, clientId, {
            metadata,
            jwks,
        }).catch((err) => {
            throw invalid_client_metadata_error_js_1.InvalidClientMetadataError.from(err, `Rejected client information for "${clientId}"`);
        });
        const isFirstParty = partialInfo?.isFirstParty ?? false;
        const isTrusted = partialInfo?.isTrusted ?? isFirstParty;
        return new client_js_1.Client(clientId, metadata, jwks, { isFirstParty, isTrusted });
    }
    async getClientMetadata(clientId) {
        if ((0, oauth_types_1.isOAuthClientIdLoopback)(clientId)) {
            return this.getLoopbackClientMetadata(clientId);
        }
        else if ((0, oauth_types_1.isOAuthClientIdDiscoverable)(clientId)) {
            return this.getDiscoverableClientMetadata(clientId);
        }
        else if (this.store) {
            return this.getStoredClientMetadata(clientId);
        }
        throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Invalid client ID "${clientId}"`);
    }
    async getLoopbackClientMetadata(clientId) {
        const { loopbackMetadata } = this;
        if (!loopbackMetadata) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Loopback clients are not allowed');
        }
        const metadata = oauth_types_1.oauthClientMetadataSchema.parse(await loopbackMetadata(clientId));
        return this.validateClientMetadata(clientId, metadata);
    }
    async getDiscoverableClientMetadata(clientId) {
        const metadataUrl = (0, client_utils_js_1.parseDiscoverableClientId)(clientId);
        const metadata = await this.metadataGetter.get(metadataUrl.href);
        // Note: we do *not* re-validate the metadata here, as the metadata is
        // validated within the getter. This is to avoid double validation.
        //
        // return this.validateClientMetadata(metadataUrl.href, metadata)
        return metadata;
    }
    async getStoredClientMetadata(clientId) {
        if (this.store) {
            const metadata = await this.store.findClient(clientId);
            return this.validateClientMetadata(clientId, metadata);
        }
        throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Invalid client ID "${clientId}"`);
    }
    /**
     * This method will ensure that the client metadata is valid w.r.t. the OAuth
     * and OIDC specifications. It will also ensure that the metadata is
     * compatible with the implementation of this library, and ATPROTO's
     * requirements.
     */
    validateClientMetadata(clientId, metadata) {
        if (metadata.jwks && metadata.jwks_uri) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('jwks_uri and jwks are mutually exclusive');
        }
        // Known OIDC specific parameters
        for (const k of [
            'default_max_age',
            'userinfo_signed_response_alg',
            'id_token_signed_response_alg',
            'userinfo_encrypted_response_alg',
        ]) {
            if (metadata[k] != null) {
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported "${k}" parameter`);
            }
        }
        const clientUriUrl = metadata.client_uri
            ? new URL(metadata.client_uri)
            : null;
        const clientUriDomain = clientUriUrl
            ? (0, hostname_js_1.parseUrlPublicSuffix)(clientUriUrl)
            : null;
        if (clientUriUrl && !clientUriDomain) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('client_uri hostname is invalid');
        }
        const scopes = metadata.scope?.split(' ');
        if (!scopes) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Missing scope property');
        }
        if (!scopes.includes('atproto')) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Missing "atproto" scope');
        }
        const dupScope = scopes?.find(isDuplicate);
        if (dupScope) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Duplicate scope "${dupScope}"`);
        }
        for (const scope of scopes) {
            // Note, once we have dynamic scopes, this check will need to be
            // updated to check against the server's supported scopes.
            if (!this.serverMetadata.scopes_supported?.includes(scope)) {
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported scope "${scope}"`);
            }
        }
        const dupGrantType = metadata.grant_types.find(isDuplicate);
        if (dupGrantType) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Duplicate grant type "${dupGrantType}"`);
        }
        for (const grantType of metadata.grant_types) {
            switch (grantType) {
                case 'implicit':
                    // Never allowed (unsafe)
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Grant type "${grantType}" is not allowed`);
                // @TODO: Add support (e.g. for first party client)
                // case 'client_credentials':
                // case 'password':
                case 'authorization_code':
                case 'refresh_token':
                    if (!this.serverMetadata.grant_types_supported?.includes(grantType)) {
                        throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported grant type "${grantType}"`);
                    }
                    break;
                default:
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Grant type "${grantType}" is not supported`);
            }
        }
        if (metadata.client_id && metadata.client_id !== clientId) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('client_id does not match');
        }
        if (metadata.subject_type && metadata.subject_type !== 'public') {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Only "public" subject_type is supported');
        }
        const method = metadata[`token_endpoint_auth_method`];
        switch (method) {
            case undefined:
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Missing token_endpoint_auth_method client metadata');
            case 'none':
                if (metadata.token_endpoint_auth_signing_alg) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`token_endpoint_auth_method "none" must not have token_endpoint_auth_signing_alg`);
                }
                break;
            case 'private_key_jwt':
                if (!metadata.jwks && !metadata.jwks_uri) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`private_key_jwt auth method requires jwks or jwks_uri`);
                }
                if (metadata.jwks?.keys.length === 0) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`private_key_jwt auth method requires at least one key in jwks`);
                }
                if (!metadata.token_endpoint_auth_signing_alg) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Missing token_endpoint_auth_signing_alg client metadata`);
                }
                break;
            default:
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`${method} is not a supported "token_endpoint_auth_method". Use "private_key_jwt" or "none".`);
        }
        if (metadata.authorization_encrypted_response_enc) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Encrypted authorization response is not supported');
        }
        if (metadata.tls_client_certificate_bound_access_tokens) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Mutual-TLS bound access tokens are not supported');
        }
        if (metadata.authorization_encrypted_response_enc &&
            !metadata.authorization_encrypted_response_alg) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('authorization_encrypted_response_enc requires authorization_encrypted_response_alg');
        }
        // ATPROTO spec requires the use of DPoP (OAuth spec defaults to false)
        if (metadata.dpop_bound_access_tokens !== true) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('"dpop_bound_access_tokens" must be true');
        }
        // ATPROTO spec requires the use of PKCE, does not support OIDC
        if (!metadata.response_types.includes('code')) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('response_types must include "code"');
        }
        else if (!metadata.grant_types.includes('authorization_code')) {
            // Consistency check
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`The "code" response type requires that "grant_types" contains "authorization_code"`);
        }
        if (metadata.authorization_details_types?.length) {
            const dupAuthDetailsType = metadata.authorization_details_types.find(isDuplicate);
            if (dupAuthDetailsType) {
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Duplicate authorization_details_type "${dupAuthDetailsType}"`);
            }
            const authorizationDetailsTypesSupported = this.serverMetadata.authorization_details_types_supported;
            if (!authorizationDetailsTypesSupported) {
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('authorization_details_types are not supported');
            }
            for (const type of metadata.authorization_details_types) {
                if (!authorizationDetailsTypesSupported.includes(type)) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported authorization_details_type "${type}"`);
                }
            }
        }
        if (!metadata.redirect_uris?.length) {
            // ATPROTO spec requires that at least one redirect URI is provided
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('At least one redirect_uri is required');
        }
        if (metadata.application_type === 'web' &&
            metadata.grant_types.includes('implicit')) {
            // https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2
            //
            // > Web Clients [as defined by "application_type"] using the OAuth
            // > Implicit Grant Type MUST only register URLs using the https
            // > scheme as redirect_uris; they MUST NOT use localhost as the
            // > hostname.
            for (const redirectUri of metadata.redirect_uris) {
                const url = (0, client_utils_js_1.parseRedirectUri)(redirectUri);
                if (url.protocol !== 'https:') {
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Web clients must use HTTPS redirect URIs`);
                }
                if (url.hostname === 'localhost') {
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Web clients must not use localhost as the hostname`);
                }
            }
        }
        for (const redirectUri of metadata.redirect_uris) {
            const url = (0, client_utils_js_1.parseRedirectUri)(redirectUri);
            if (url.username || url.password) {
                // Is this a valid concern? Should we allow credentials in the URI?
                throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Redirect URI ${url} must not contain credentials`);
            }
            switch (true) {
                // FIRST: Loopback redirect URI exception (only for native apps)
                case url.hostname === 'localhost': {
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-8.3
                    //
                    // > While redirect URIs using localhost (i.e.,
                    // > "http://localhost:{port}/{path}") function similarly to loopback IP
                    // > redirects described in Section 7.3, the use of localhost is NOT
                    // > RECOMMENDED. Specifying a redirect URI with the loopback IP literal
                    // > rather than localhost avoids inadvertently listening on network
                    // > interfaces other than the loopback interface. It is also less
                    // > susceptible to client-side firewalls and misconfigured host name
                    // > resolution on the user's device.
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Loopback redirect URI ${url} is not allowed (use explicit IPs instead)`);
                }
                case url.hostname === '127.0.0.1':
                case url.hostname === '[::1]': {
                    // Only allowed for native apps
                    if (metadata.application_type !== 'native') {
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Loopback redirect URIs are only allowed for native apps`);
                    }
                    if (url.port) {
                        // https://datatracker.ietf.org/doc/html/rfc8252#section-7.3
                        //
                        // > The authorization server MUST allow any port to be specified at
                        // > the time of the request for loopback IP redirect URIs, to
                        // > accommodate clients that obtain an available ephemeral port
                        // > from the operating system at the time of the request.
                        //
                        // Note: although validation of the redirect_uri will ignore the
                        // port we still allow it to be specified, as the spec does not
                        // forbid it. If a port number is specified, ports will need to
                        // match when validating authorization requests. See
                        // "compareRedirectUri()".
                    }
                    if (url.protocol !== 'http:') {
                        // https://datatracker.ietf.org/doc/html/rfc8252#section-7.3
                        //
                        // > Loopback redirect URIs use the "http" scheme and are constructed
                        // > with the loopback IP literal and whatever port the client is
                        // > listening on. That is, "http://127.0.0.1:{port}/{path}" for IPv4,
                        // > and "http://[::1]:{port}/{path}" for IPv6.
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Loopback redirect URI ${url} must use HTTP`);
                    }
                    break;
                }
                // SECOND: Protocol-based URI Redirection
                case url.protocol === 'http:': {
                    // https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2
                    //
                    // > request_uri [...] URLs MUST use the https scheme unless the
                    // > target Request Object is signed in a way that is verifiable by
                    // > the OP.
                    //
                    // OIDC/Request Object are not supported. ATproto spec should not
                    // allow HTTP redirect URIs either.
                    // https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2
                    //
                    // > Authorization Servers MAY reject Redirection URI values using
                    // > the http scheme, other than the loopback case for Native
                    // > Clients.
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError('Only loopback redirect URIs are allowed to use the "http" scheme');
                }
                case url.protocol === 'https:': {
                    if (!(0, hostname_js_1.isInternetUrl)(url)) {
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Redirect URI "${url}"'s domain name must belong to the Public Suffix List (PSL)`);
                    }
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-8.4
                    //
                    // > In addition to the collision-resistant properties, requiring a
                    // > URI scheme based on a domain name that is under the control of
                    // > the app can help to prove ownership in the event of a dispute
                    // > where two apps claim the same private-use URI scheme (where one
                    // > app is acting maliciously).
                    //
                    // We can't enforce this here (in generic client validation) because
                    // we don't have a concept of generic proven ownership.
                    //
                    // Discoverable clients, however, will have this check covered in the
                    // `validateDiscoverableClientMetadata`, by using the client_id's
                    // domain as "proven ownership".
                    // The following restriction from OIDC is *not* enforced for clients
                    // as it prevents "App Links" / "Apple Universal Links" from being
                    // used as redirect URIs.
                    //
                    // https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2
                    //
                    // > Native Clients [as defined by "application_type"] MUST only
                    // > register redirect_uris using custom URI schemes or loopback URLs
                    // > using the http scheme; loopback URLs use localhost or the IP
                    // > loopback literals 127.0.0.1 or [::1] as the hostname.
                    //
                    // if (metadata.application_type === 'native') {
                    //   throw new InvalidRedirectUriError(
                    //     `Native clients must use custom URI schemes or loopback URLs`,
                    //   )
                    // }
                    break;
                }
                case isPrivateUseUriScheme(url): {
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-7.1
                    //
                    // > When choosing a URI scheme to associate with the app, apps MUST
                    // > use a URI scheme based on a domain name under their control,
                    // > expressed in reverse order, as recommended by Section 3.8 of
                    // > [RFC7595] for private-use URI schemes.
                    if (metadata.application_type !== 'native') {
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Private-Use URI Scheme redirect URI are only allowed for native apps`);
                    }
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-8.4
                    //
                    // > In addition to the collision-resistant properties, requiring a
                    // > URI scheme based on a domain name that is under the control of
                    // > the app can help to prove ownership in the event of a dispute
                    // > where two apps claim the same private-use URI scheme (where one
                    // > app is acting maliciously).
                    //
                    // We can't check for ownership here (as there is no concept of
                    // proven ownership in the generic client validation), but we can
                    // check that the domain is a valid domain name.
                    const urlDomain = reverseDomain(url.protocol.slice(0, -1));
                    if (!(0, hostname_js_1.isInternetHost)(urlDomain)) {
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Private-use URI Scheme redirect URI must be based on a valid domain name`);
                    }
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-7.1
                    //
                    // > Following the requirements of Section 3.2 of [RFC3986], as there
                    // > is no naming authority for private-use URI scheme redirects, only
                    // > a single slash ("/") appears after the scheme component.
                    if (url.href.startsWith(`${url.protocol}//`) ||
                        url.username ||
                        url.password ||
                        url.hostname ||
                        url.port) {
                        throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Private-Use URI Scheme must be in the form ${url.protocol}/<path>`);
                    }
                    break;
                }
                default:
                    // https://datatracker.ietf.org/doc/html/rfc8252#section-8.4
                    //
                    // > At a minimum, any private-use URI scheme that doesn't contain a
                    // > period character (".") SHOULD be rejected.
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Invalid redirect URI scheme "${url.protocol}"`);
            }
        }
        if ((0, oauth_types_1.isOAuthClientIdLoopback)(clientId)) {
            return this.validateLoopbackClientMetadata(clientId, metadata);
        }
        else if ((0, oauth_types_1.isOAuthClientIdDiscoverable)(clientId)) {
            return this.validateDiscoverableClientMetadata(clientId, metadata);
        }
        else {
            return metadata;
        }
    }
    validateLoopbackClientMetadata(clientId, metadata) {
        if (metadata.client_uri) {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('client_uri is not allowed for loopback clients');
        }
        if (metadata.application_type !== 'native') {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError('Loopback clients must have application_type "native"');
        }
        const method = metadata[`token_endpoint_auth_method`];
        if (method !== 'none') {
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Loopback clients are not allowed to use "token_endpoint_auth_method" ${method}`);
        }
        for (const redirectUri of metadata.redirect_uris) {
            const url = (0, client_utils_js_1.parseRedirectUri)(redirectUri);
            if (url.protocol !== 'http:') {
                throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Loopback clients must use HTTP redirect URIs`);
            }
            if (!(0, oauth_types_1.isLoopbackHost)(url.hostname)) {
                throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Loopback clients must use loopback redirect URIs`);
            }
        }
        return metadata;
    }
    validateDiscoverableClientMetadata(clientId, metadata) {
        if (!metadata.client_id) {
            // https://drafts.aaronpk.com/draft-parecki-oauth-client-id-metadata-document/draft-parecki-oauth-client-id-metadata-document.html
            throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`client_id is required for discoverable clients`);
        }
        const clientIdUrl = (0, client_utils_js_1.parseDiscoverableClientId)(clientId);
        if (metadata.client_uri) {
            // https://drafts.aaronpk.com/draft-parecki-oauth-client-id-metadata-document/draft-parecki-oauth-client-id-metadata-document.html
            //
            // The client_uri must be a parent of the client_id URL. This might be
            // relaxed in the future.
            const clientUriUrl = new URL(metadata.client_uri);
            if (clientUriUrl.origin !== clientIdUrl.origin) {
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`client_uri must have the same origin as the client_id`);
            }
            if (clientIdUrl.pathname !== clientUriUrl.pathname) {
                if (!clientIdUrl.pathname.startsWith(clientUriUrl.pathname.endsWith('/')
                    ? clientUriUrl.pathname
                    : `${clientUriUrl.pathname}/`)) {
                    throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`client_uri must be a parent URL of the client_id`);
                }
            }
        }
        const method = metadata[`token_endpoint_auth_method`];
        switch (method) {
            case 'none':
            case 'private_key_jwt':
            case undefined:
                break;
            case 'client_secret_post':
            case 'client_secret_basic':
            case 'client_secret_jwt':
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Client authentication method "${method}" is not allowed for discoverable clients`);
            default:
                throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported client authentication method "${method}"`);
        }
        for (const redirectUri of metadata.redirect_uris) {
            const url = (0, client_utils_js_1.parseRedirectUri)(redirectUri);
            if (isPrivateUseUriScheme(url)) {
                // https://datatracker.ietf.org/doc/html/rfc8252#section-8.4
                //
                // > In addition to the collision-resistant properties, requiring a
                // > URI scheme based on a domain name that is under the control of
                // > the app can help to prove ownership in the event of a dispute
                // > where two apps claim the same private-use URI scheme (where one
                // > app is acting maliciously).
                // https://drafts.aaronpk.com/draft-parecki-oauth-client-id-metadata-document/draft-parecki-oauth-client-id-metadata-document.html
                //
                // Fully qualified domain name (FQDN) of the client_id, in reverse
                // order. This could be relaxed to allow same apex domain names, or
                // parent domains, but for now we require an exact match.
                const protocol = `${reverseDomain(clientIdUrl.hostname)}:`;
                if (url.protocol !== protocol) {
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Private-Use URI Scheme redirect URI, for discoverable client metadata, must be the fully qualified domain name (FQDN) of the client_id, in reverse order (${protocol})`);
                }
            }
            if (url.protocol === 'https:') {
                // https://datatracker.ietf.org/doc/html/rfc8252#section-8.4
                //
                // > In addition to the collision-resistant properties, requiring a
                // > URI scheme based on a domain name that is under the control of
                // > the app can help to prove ownership in the event of a dispute
                // > where two apps claim the same private-use URI scheme (where one
                // > app is acting maliciously).
                //
                // Although this only applies to "native" clients (extract being from
                // rfc8252), we apply this rule to "web" clients as well.
                if (url.hostname !== clientIdUrl.hostname) {
                    throw new invalid_redirect_uri_error_js_1.InvalidRedirectUriError(`Redirect URI ${url} must be under the same domain as client_id ${metadata.client_uri}`);
                }
            }
        }
        return metadata;
    }
}
exports.ClientManager = ClientManager;
function isDuplicate(value, index, array) {
    return array.includes(value, index + 1);
}
function reverseDomain(domain) {
    return domain.split('.').reverse().join('.');
}
function isPrivateUseUriScheme(uri) {
    return uri.protocol.includes('.');
}
function buildJsonGetRequest(uri, options) {
    return new Request(uri, {
        headers: { accept: 'application/json' },
        // @ts-expect-error invalid types in "undici-types"
        cache: options?.noCache ? 'no-cache' : undefined,
        signal: options?.signal,
        redirect: 'error',
    });
}
