// @atproto/oauth-provider/dist/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const jose_1 = require("jose");
const oauth_types_1 = require("@atproto/oauth-types");
const constants_js_1 = require("../constants.js");
const invalid_authorization_details_error_js_1 = require("../errors/invalid-authorization-details-error.js");
const invalid_client_error_js_1 = require("../errors/invalid-client-error.js");
const invalid_client_metadata_error_js_1 = require("../errors/invalid-client-metadata-error.js");
const invalid_parameters_error_js_1 = require("../errors/invalid-parameters-error.js");
const invalid_request_error_js_1 = require("../errors/invalid-request-error.js");
const invalid_scope_error_js_1 = require("../errors/invalid-scope-error.js");
const redirect_uri_js_1 = require("../lib/util/redirect-uri.js");
const client_auth_js_1 = require("./client-auth.js");
const { JOSEError } = jose_1.errors;
class Client {
    id;
    metadata;
    jwks;
    info;
    /**
     * @see {@link https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#token-endpoint-auth-method}
     */
    static AUTH_METHODS_SUPPORTED = ['none', 'private_key_jwt'];
    keyGetter;
    constructor(id, metadata, jwks = metadata.jwks, info) {
        this.id = id;
        this.metadata = metadata;
        this.jwks = jwks;
        this.info = info;
        // If the remote JWKS content is provided, we don't need to fetch it again.
        this.keyGetter =
            jwks || !metadata.jwks_uri
                ? (0, jose_1.createLocalJWKSet)(jwks || { keys: [] })
                : (0, jose_1.createRemoteJWKSet)(new URL(metadata.jwks_uri), {});
    }
    async decodeRequestObject(jar) {
        try {
            switch (this.metadata.request_object_signing_alg) {
                case 'none':
                    return await this.jwtVerifyUnsecured(jar, {
                        maxTokenAge: constants_js_1.JAR_MAX_AGE / 1000,
                    });
                case undefined:
                    // https://openid.net/specs/openid-connect-registration-1_0.html#rfc.section.2
                    // > The default, if omitted, is that any algorithm supported by the OP
                    // > and the RP MAY be used.
                    return await this.jwtVerify(jar, {
                        maxTokenAge: constants_js_1.JAR_MAX_AGE / 1000,
                    });
                default:
                    return await this.jwtVerify(jar, {
                        maxTokenAge: constants_js_1.JAR_MAX_AGE / 1000,
                        algorithms: [this.metadata.request_object_signing_alg],
                    });
            }
        }
        catch (err) {
            const message = err instanceof JOSEError
                ? `Invalid "request" object: ${err.message}`
                : `Invalid "request" object`;
            throw new invalid_request_error_js_1.InvalidRequestError(message, err);
        }
    }
    async jwtVerifyUnsecured(token, options) {
        return jose_1.UnsecuredJWT.decode(token, {
            ...options,
            issuer: this.id,
        });
    }
    async jwtVerify(token, options) {
        return (0, jose_1.jwtVerify)(token, this.keyGetter, {
            ...options,
            issuer: this.id,
        });
    }
    /**
     * @see {@link https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1}
     * @see {@link https://datatracker.ietf.org/doc/html/rfc7523#section-3}
     * @see {@link https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#token-endpoint-auth-method}
     */
    async verifyCredentials(input, checks) {
        const method = this.metadata[`token_endpoint_auth_method`];
        if (method === 'none') {
            const clientAuth = { method: 'none' };
            return { clientAuth };
        }
        if (method === 'private_key_jwt') {
            if (!('client_assertion_type' in input) || !input.client_assertion_type) {
                throw new invalid_request_error_js_1.InvalidRequestError(`client_assertion_type required for "${method}"`);
            }
            else if (!input.client_assertion) {
                throw new invalid_request_error_js_1.InvalidRequestError(`client_assertion required for "${method}"`);
            }
            if (input.client_assertion_type === oauth_types_1.CLIENT_ASSERTION_TYPE_JWT_BEARER) {
                const result = await this.jwtVerify(input.client_assertion, {
                    audience: checks.audience,
                    subject: this.id,
                    maxTokenAge: constants_js_1.CLIENT_ASSERTION_MAX_AGE / 1000,
                    requiredClaims: ['jti'],
                }).catch((err) => {
                    if (err instanceof JOSEError) {
                        const msg = `Validation of "client_assertion" failed: ${err.message}`;
                        throw new invalid_client_error_js_1.InvalidClientError(msg, err);
                    }
                    throw err;
                });
                if (!result.protectedHeader.kid) {
                    throw new invalid_client_error_js_1.InvalidClientError(`"kid" required in client_assertion`);
                }
                const clientAuth = {
                    method: oauth_types_1.CLIENT_ASSERTION_TYPE_JWT_BEARER,
                    jkt: await (0, client_auth_js_1.authJwkThumbprint)(result.key),
                    alg: result.protectedHeader.alg,
                    kid: result.protectedHeader.kid,
                };
                return { clientAuth, nonce: result.payload.jti };
            }
            throw new invalid_client_error_js_1.InvalidClientError(`Unsupported client_assertion_type "${input.client_assertion_type}"`);
        }
        // @ts-expect-error Ensure to keep Client.AUTH_METHODS_SUPPORTED in sync
        // with the implementation of this function.
        if (Client.AUTH_METHODS_SUPPORTED.includes(method)) {
            throw new Error(`verifyCredentials() should implement all of ${[
                Client.AUTH_METHODS_SUPPORTED,
            ]}`);
        }
        throw new invalid_client_metadata_error_js_1.InvalidClientMetadataError(`Unsupported token_endpoint_auth_method "${method}"`);
    }
    /**
     * Ensures that a {@link ClientAuth} generated in the past is still valid wrt
     * the current client metadata & jwks. This is used to invalidate tokens when
     * the client stops advertising the key that it used to authenticate itself
     * during the initial token request.
     */
    async validateClientAuth(clientAuth) {
        if (clientAuth.method === 'none') {
            return this.metadata[`token_endpoint_auth_method`] === 'none';
        }
        if (clientAuth.method === oauth_types_1.CLIENT_ASSERTION_TYPE_JWT_BEARER) {
            if (this.metadata[`token_endpoint_auth_method`] !== 'private_key_jwt') {
                return false;
            }
            try {
                const key = await this.keyGetter({
                    kid: clientAuth.kid,
                    alg: clientAuth.alg,
                }, { payload: '', signature: '' });
                const jtk = await (0, client_auth_js_1.authJwkThumbprint)(key);
                return jtk === clientAuth.jkt;
            }
            catch (e) {
                return false;
            }
        }
        // @ts-expect-error
        throw new Error(`Invalid method "${clientAuth.method}"`);
    }
    /**
     * Validates the request parameters against the client metadata.
     */
    validateRequest(parameters) {
        if (parameters.client_id !== this.id) {
            throw new invalid_parameters_error_js_1.InvalidParametersError(parameters, 'The "client_id" parameter field does not match the value used to authenticate the client');
        }
        if (parameters.scope !== undefined) {
            // Any scope requested by the client must be registered in the client
            // metadata.
            const declaredScopes = this.metadata.scope?.split(' ');
            if (!declaredScopes) {
                throw new invalid_scope_error_js_1.InvalidScopeError(parameters, 'Client has no declared scopes in its metadata');
            }
            for (const scope of parameters.scope.split(' ')) {
                if (!declaredScopes.includes(scope)) {
                    throw new invalid_scope_error_js_1.InvalidScopeError(parameters, `Scope "${scope}" is not declared in the client metadata`);
                }
            }
        }
        if (!this.metadata.response_types.includes(parameters.response_type)) {
            throw new invalid_parameters_error_js_1.InvalidParametersError(parameters, `Invalid response_type "${parameters.response_type}" requested by the client`);
        }
        if (parameters.response_type.includes('code')) {
            if (!this.metadata.grant_types.includes('authorization_code')) {
                throw new invalid_parameters_error_js_1.InvalidParametersError(parameters, `This client is not allowed to use the "authorization_code" grant type`);
            }
        }
        const { redirect_uri } = parameters;
        if (redirect_uri) {
            if (!this.metadata.redirect_uris.some((uri) => (0, redirect_uri_js_1.compareRedirectUri)(uri, redirect_uri))) {
                throw new invalid_parameters_error_js_1.InvalidParametersError(parameters, `Invalid redirect_uri ${redirect_uri}`);
            }
        }
        else {
            const { defaultRedirectUri } = this;
            if (defaultRedirectUri) {
                parameters = { ...parameters, redirect_uri: defaultRedirectUri };
            }
            else {
                // https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-10#authorization-request
                //
                // > "redirect_uri": OPTIONAL if only one redirect URI is registered for
                // > this client. REQUIRED if multiple redirect URIs are registered for this
                // > client.
                throw new invalid_parameters_error_js_1.InvalidParametersError(parameters, 'redirect_uri is required');
            }
        }
        if (parameters.authorization_details) {
            const { authorization_details_types } = this.metadata;
            if (!authorization_details_types) {
                throw new invalid_authorization_details_error_js_1.InvalidAuthorizationDetailsError(parameters, 'Client Metadata does not declare any "authorization_details"');
            }
            for (const detail of parameters.authorization_details) {
                if (!authorization_details_types?.includes(detail.type)) {
                    throw new invalid_authorization_details_error_js_1.InvalidAuthorizationDetailsError(parameters, `Client Metadata does not declare any "authorization_details" of type "${detail.type}"`);
                }
            }
        }
        return parameters;
    }
    get defaultRedirectUri() {
        const { redirect_uris } = this.metadata;
        return redirect_uris.length === 1 ? redirect_uris[0] : undefined;
    }
}
exports.Client = Client;
