// @atproto/oauth-provider/dist/token/token-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenManager = void 0;
const node_crypto_1 = require("node:crypto");
const jwk_1 = require("@atproto/jwk");
const oauth_types_1 = require("@atproto/oauth-types");
const access_token_type_js_1 = require("../access-token/access-token-type.js");
const constants_js_1 = require("../constants.js");
const invalid_dpop_key_binding_error_js_1 = require("../errors/invalid-dpop-key-binding-error.js");
const invalid_dpop_proof_error_js_1 = require("../errors/invalid-dpop-proof-error.js");
const invalid_grant_error_js_1 = require("../errors/invalid-grant-error.js");
const invalid_request_error_js_1 = require("../errors/invalid-request-error.js");
const invalid_token_error_js_1 = require("../errors/invalid-token-error.js");
const date_js_1 = require("../lib/util/date.js");
const function_js_1 = require("../lib/util/function.js");
const code_js_1 = require("../request/code.js");
const refresh_token_js_1 = require("./refresh-token.js");
const token_id_js_1 = require("./token-id.js");
const verify_token_claims_js_1 = require("./verify-token-claims.js");
class TokenManager {
    store;
    signer;
    hooks;
    accessTokenType;
    tokenMaxAge;
    constructor(store, signer, hooks, accessTokenType, tokenMaxAge = constants_js_1.TOKEN_MAX_AGE) {
        this.store = store;
        this.signer = signer;
        this.hooks = hooks;
        this.accessTokenType = accessTokenType;
        this.tokenMaxAge = tokenMaxAge;
    }
    createTokenExpiry(now = new Date()) {
        return new Date(now.getTime() + this.tokenMaxAge);
    }
    useJwtAccessToken(account) {
        if (this.accessTokenType === access_token_type_js_1.AccessTokenType.auto) {
            return this.signer.issuer !== account.aud;
        }
        return this.accessTokenType === access_token_type_js_1.AccessTokenType.jwt;
    }
    async create(client, clientAuth, clientMetadata, account, device, parameters, input, dpopJkt) {
        // @NOTE the atproto specific DPoP requirement is enforced though the
        // "dpop_bound_access_tokens" metadata, which is enforced by the
        // ClientManager class.
        if (client.metadata.dpop_bound_access_tokens && !dpopJkt) {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP proof required');
        }
        if (!parameters.dpop_jkt) {
            // Allow clients to bind their access tokens to a DPoP key during
            // token request if they didn't provide a "dpop_jkt" during the
            // authorization request.
            if (dpopJkt)
                parameters = { ...parameters, dpop_jkt: dpopJkt };
        }
        else if (parameters.dpop_jkt !== dpopJkt) {
            throw new invalid_dpop_key_binding_error_js_1.InvalidDpopKeyBindingError();
        }
        if (clientAuth.method === oauth_types_1.CLIENT_ASSERTION_TYPE_JWT_BEARER) {
            // Clients **must not** use their private key to sign DPoP proofs.
            if (parameters.dpop_jkt && clientAuth.jkt === parameters.dpop_jkt) {
                throw new invalid_request_error_js_1.InvalidRequestError('The DPoP proof must be signed with a different key than the client assertion');
            }
        }
        if (!client.metadata.grant_types.includes(input.grant_type)) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`This client is not allowed to use the "${input.grant_type}" grant type`);
        }
        let code = null;
        switch (input.grant_type) {
            case 'authorization_code': {
                if (!(0, code_js_1.isCode)(input.code)) {
                    throw new invalid_grant_error_js_1.InvalidGrantError('Invalid code');
                }
                const tokenInfo = await this.store.findTokenByCode(input.code);
                if (tokenInfo) {
                    await this.store.deleteToken(tokenInfo.id);
                    throw new invalid_grant_error_js_1.InvalidGrantError(`Code replayed`);
                }
                code = input.code;
                if (parameters.redirect_uri !== input.redirect_uri) {
                    throw new invalid_grant_error_js_1.InvalidGrantError('The redirect_uri parameter must match the one used in the authorization request');
                }
                if (parameters.code_challenge) {
                    if (!input.code_verifier) {
                        throw new invalid_grant_error_js_1.InvalidGrantError('code_verifier is required');
                    }
                    if (input.code_verifier.length < 43) {
                        throw new invalid_grant_error_js_1.InvalidGrantError('code_verifier too short');
                    }
                    switch (parameters.code_challenge_method ?? 'plain') {
                        case 'plain': {
                            if (parameters.code_challenge !== input.code_verifier) {
                                throw new invalid_grant_error_js_1.InvalidGrantError('Invalid code_verifier');
                            }
                            break;
                        }
                        case 'S256': {
                            const inputChallenge = Buffer.from(parameters.code_challenge, 'base64');
                            const computedChallenge = (0, node_crypto_1.createHash)('sha256')
                                .update(input.code_verifier)
                                .digest();
                            if (inputChallenge.compare(computedChallenge) !== 0) {
                                throw new invalid_grant_error_js_1.InvalidGrantError('Invalid code_verifier');
                            }
                            break;
                        }
                        default: {
                            // Should never happen (because request validation should catch this)
                            throw new Error(`Unsupported code_challenge_method`);
                        }
                    }
                }
                else if (input.code_verifier !== undefined) {
                    throw new invalid_request_error_js_1.InvalidRequestError("code_challenge parameter wasn't provided");
                }
                if (!device) {
                    // Fool-proofing (authorization_code grant should always have a device)
                    throw new invalid_request_error_js_1.InvalidRequestError('consent was not given for this device');
                }
                break;
            }
            default: {
                // Other grants (e.g "password", "client_credentials") could be added
                // here in the future...
                throw new invalid_request_error_js_1.InvalidRequestError(`Unsupported grant type "${input.grant_type}"`);
            }
        }
        const tokenId = await (0, token_id_js_1.generateTokenId)();
        const refreshToken = client.metadata.grant_types.includes('refresh_token')
            ? await (0, refresh_token_js_1.generateRefreshToken)()
            : undefined;
        const now = new Date();
        const expiresAt = this.createTokenExpiry(now);
        const authorizationDetails = await (0, function_js_1.callAsync)(this.hooks.getAuthorizationDetails, {
            client,
            clientAuth,
            clientMetadata,
            parameters,
            account,
        });
        const tokenData = {
            createdAt: now,
            updatedAt: now,
            expiresAt,
            clientId: client.id,
            clientAuth,
            deviceId: device?.id ?? null,
            sub: account.sub,
            parameters,
            details: authorizationDetails ?? null,
            code,
        };
        await this.store.createToken(tokenId, tokenData, refreshToken);
        try {
            const accessToken = !this.useJwtAccessToken(account)
                ? tokenId
                : await this.signer.accessToken(client, parameters, {
                    // We don't specify the alg here. We suppose the Resource server will be
                    // able to verify the token using any alg.
                    aud: account.aud,
                    sub: account.sub,
                    alg: undefined,
                    exp: expiresAt,
                    iat: now,
                    jti: tokenId,
                    cnf: parameters.dpop_jkt ? { jkt: parameters.dpop_jkt } : undefined,
                    authorization_details: authorizationDetails,
                });
            const response = await this.buildTokenResponse(client, accessToken, refreshToken, expiresAt, parameters, account, authorizationDetails);
            await (0, function_js_1.callAsync)(this.hooks.onTokenCreated, {
                client,
                clientAuth,
                clientMetadata,
                account,
                parameters,
                deviceId: device ? device.id : null,
            });
            return response;
        }
        catch (err) {
            // Just in case the token could not be issued, we delete it from the store
            await this.store.deleteToken(tokenId);
            throw err;
        }
    }
    async buildTokenResponse(client, accessToken, refreshToken, expiresAt, parameters, account, authorizationDetails) {
        const tokenResponse = {
            access_token: accessToken,
            token_type: parameters.dpop_jkt ? 'DPoP' : 'Bearer',
            refresh_token: refreshToken,
            scope: parameters.scope,
            authorization_details: authorizationDetails,
            get expires_in() {
                return (0, date_js_1.dateToRelativeSeconds)(expiresAt);
            },
            // ATPROTO extension: add the sub claim to the token response to allow
            // clients to resolve the PDS url (audience) using the did resolution
            // mechanism.
            sub: account.sub,
        };
        return tokenResponse;
    }
    async validateAccess(client, clientAuth, tokenInfo) {
        if (tokenInfo.data.clientId !== client.id) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Token was not issued to this client`);
        }
        if (tokenInfo.info?.authorizedClients.includes(client.id) === false) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Client no longer trusted by user`);
        }
        if (tokenInfo.data.clientAuth.method !== clientAuth.method) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Client authentication method mismatch`);
        }
        if (!(await client.validateClientAuth(tokenInfo.data.clientAuth))) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Client authentication mismatch`);
        }
    }
    async refresh(client, clientAuth, clientMetadata, input, dpopJkt) {
        const refreshTokenParsed = refresh_token_js_1.refreshTokenSchema.safeParse(input.refresh_token);
        if (!refreshTokenParsed.success) {
            throw new invalid_request_error_js_1.InvalidRequestError('Invalid refresh token');
        }
        const refreshToken = refreshTokenParsed.data;
        const tokenInfo = await this.store.findTokenByRefreshToken(refreshToken);
        if (!tokenInfo?.currentRefreshToken) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Invalid refresh token`);
        }
        const { account, data } = tokenInfo;
        const { parameters } = data;
        try {
            if (tokenInfo.currentRefreshToken !== refreshToken) {
                throw new invalid_grant_error_js_1.InvalidGrantError(`refresh token replayed`);
            }
            await this.validateAccess(client, clientAuth, tokenInfo);
            if (input.grant_type !== 'refresh_token') {
                // Fool-proofing (should never happen)
                throw new invalid_grant_error_js_1.InvalidGrantError(`Invalid grant type`);
            }
            if (!client.metadata.grant_types.includes(input.grant_type)) {
                // In case the client metadata was updated after the token was issued
                throw new invalid_grant_error_js_1.InvalidGrantError(`This client is not allowed to use the "${input.grant_type}" grant type`);
            }
            if (parameters.dpop_jkt) {
                if (!dpopJkt) {
                    throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP proof required');
                }
                else if (parameters.dpop_jkt !== dpopJkt) {
                    throw new invalid_dpop_key_binding_error_js_1.InvalidDpopKeyBindingError();
                }
            }
            const lastActivity = data.updatedAt;
            const inactivityTimeout = clientAuth.method === 'none' && !client.info.isFirstParty
                ? constants_js_1.UNAUTHENTICATED_REFRESH_INACTIVITY_TIMEOUT
                : constants_js_1.AUTHENTICATED_REFRESH_INACTIVITY_TIMEOUT;
            if (lastActivity.getTime() + inactivityTimeout < Date.now()) {
                throw new invalid_grant_error_js_1.InvalidGrantError(`Refresh token exceeded inactivity timeout`);
            }
            const lifetime = clientAuth.method === 'none' && !client.info.isFirstParty
                ? constants_js_1.UNAUTHENTICATED_REFRESH_LIFETIME
                : constants_js_1.AUTHENTICATED_REFRESH_LIFETIME;
            if (data.createdAt.getTime() + lifetime < Date.now()) {
                throw new invalid_grant_error_js_1.InvalidGrantError(`Refresh token expired`);
            }
            const authorizationDetails = await (0, function_js_1.callAsync)(this.hooks.getAuthorizationDetails, {
                client,
                clientAuth,
                clientMetadata,
                parameters,
                account,
            });
            const nextTokenId = await (0, token_id_js_1.generateTokenId)();
            const nextRefreshToken = await (0, refresh_token_js_1.generateRefreshToken)();
            const now = new Date();
            const expiresAt = this.createTokenExpiry(now);
            await this.store.rotateToken(tokenInfo.id, nextTokenId, nextRefreshToken, {
                updatedAt: now,
                expiresAt,
                // When clients rotate their public keys, we store the key that was
                // used by the client to authenticate itself while requesting new
                // tokens. The validateAccess() method will ensure that the client
                // still advertises the key that was used to issue the previous
                // refresh token. If a client stops advertising a key, all tokens
                // bound to that key will no longer be be refreshable. This allows
                // clients to proactively invalidate tokens when a key is compromised.
                // Note that the original DPoP key cannot be rotated. This protects
                // users in case the ownership of the client id changes. In the latter
                // case, a malicious actor could still advertises the public keys of
                // the previous owner, but the new owner would not be able to present
                // a valid DPoP proof.
                clientAuth,
            });
            const accessToken = !this.useJwtAccessToken(account)
                ? nextTokenId
                : await this.signer.accessToken(client, parameters, {
                    // We don't specify the alg here. We suppose the Resource server will be
                    // able to verify the token using any alg.
                    aud: account.aud,
                    sub: account.sub,
                    alg: undefined,
                    exp: expiresAt,
                    iat: now,
                    jti: nextTokenId,
                    cnf: parameters.dpop_jkt ? { jkt: parameters.dpop_jkt } : undefined,
                    authorization_details: authorizationDetails,
                });
            const response = await this.buildTokenResponse(client, accessToken, nextRefreshToken, expiresAt, parameters, account, authorizationDetails);
            await (0, function_js_1.callAsync)(this.hooks.onTokenRefreshed, {
                client,
                clientAuth,
                clientMetadata,
                account,
                parameters,
                deviceId: tokenInfo.data.deviceId,
            });
            return response;
        }
        catch (err) {
            // Just in case the token could not be refreshed, we delete it from the store
            await this.store.deleteToken(tokenInfo.id);
            throw err;
        }
    }
    /**
     * @see {@link https://datatracker.ietf.org/doc/html/rfc7009#section-2.2 | RFC7009 Section 2.2}
     */
    async revoke(token) {
        switch (true) {
            case (0, token_id_js_1.isTokenId)(token): {
                await this.store.deleteToken(token);
                return;
            }
            case (0, jwk_1.isSignedJwt)(token): {
                const { payload } = await this.signer.verify(token, {
                    clockTolerance: Infinity,
                    requiredClaims: ['jti'],
                });
                const tokenId = token_id_js_1.tokenIdSchema.parse(payload.jti);
                await this.store.deleteToken(tokenId);
                return;
            }
            case (0, refresh_token_js_1.isRefreshToken)(token): {
                const tokenInfo = await this.store.findTokenByRefreshToken(token);
                if (tokenInfo)
                    await this.store.deleteToken(tokenInfo.id);
                return;
            }
            case (0, code_js_1.isCode)(token): {
                const tokenInfo = await this.store.findTokenByCode(token);
                if (tokenInfo)
                    await this.store.deleteToken(tokenInfo.id);
                return;
            }
            default:
                // No error should be returned if the token is not valid
                return;
        }
    }
    /**
     * Allows an (authenticated) client to obtain information about a token.
     *
     * @see {@link https://datatracker.ietf.org/doc/html/rfc7662 RFC7662}
     */
    async clientTokenInfo(client, clientAuth, token) {
        const tokenInfo = await this.findTokenInfo(token);
        if (!tokenInfo) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Invalid token`);
        }
        try {
            await this.validateAccess(client, clientAuth, tokenInfo);
        }
        catch (err) {
            await this.store.deleteToken(tokenInfo.id);
            throw err;
        }
        if (tokenInfo.data.expiresAt.getTime() < Date.now()) {
            throw new invalid_grant_error_js_1.InvalidGrantError(`Token expired`);
        }
        return tokenInfo;
    }
    async findTokenInfo(token) {
        switch (true) {
            case (0, token_id_js_1.isTokenId)(token):
                return this.store.readToken(token);
            case (0, jwk_1.isSignedJwt)(token): {
                const { payload } = await this.signer
                    .verifyAccessToken(token)
                    .catch((_) => ({ payload: null }));
                if (!payload)
                    return null;
                const tokenInfo = await this.store.readToken(payload.jti);
                if (!tokenInfo)
                    return null;
                // Audience changed (e.g. user was moved to another resource server)
                if (payload.aud !== tokenInfo.account.aud) {
                    return null;
                }
                // Invalid store implementation ?
                if (payload.sub !== tokenInfo.account.sub) {
                    throw new Error(`Account sub (${tokenInfo.account.sub}) does not match token sub (${payload.sub})`);
                }
                return tokenInfo;
            }
            case (0, refresh_token_js_1.isRefreshToken)(token): {
                const tokenInfo = await this.store.findTokenByRefreshToken(token);
                if (!tokenInfo?.currentRefreshToken)
                    return null;
                if (tokenInfo.currentRefreshToken !== token)
                    return null;
                return tokenInfo;
            }
            default:
                // Should never happen
                return null;
        }
    }
    async getTokenInfo(tokenType, tokenId) {
        const tokenInfo = await this.store.readToken(tokenId);
        if (!tokenInfo) {
            throw new invalid_token_error_js_1.InvalidTokenError(tokenType, `Invalid token`);
        }
        if (!(tokenInfo.data.expiresAt.getTime() > Date.now())) {
            throw new invalid_token_error_js_1.InvalidTokenError(tokenType, `Token expired`);
        }
        return tokenInfo;
    }
    async authenticateTokenId(tokenType, token, dpopJkt, verifyOptions) {
        const tokenInfo = await this.getTokenInfo(tokenType, token);
        const { parameters } = tokenInfo.data;
        // Construct a list of claim, as if the token was a JWT.
        const claims = {
            aud: tokenInfo.account.aud,
            sub: tokenInfo.account.sub,
            exp: (0, date_js_1.dateToEpoch)(tokenInfo.data.expiresAt),
            iat: (0, date_js_1.dateToEpoch)(tokenInfo.data.updatedAt),
            scope: tokenInfo.data.parameters.scope,
            client_id: tokenInfo.data.clientId,
            cnf: parameters.dpop_jkt ? { jkt: parameters.dpop_jkt } : undefined,
        };
        const result = (0, verify_token_claims_js_1.verifyTokenClaims)(token, token, tokenType, dpopJkt, claims, verifyOptions);
        return { ...result, tokenInfo };
    }
}
exports.TokenManager = TokenManager;
