// @atproto/oauth-provider/dist/lib/http/request.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateHeaderValue = validateHeaderValue;
exports.validateFetchMode = validateFetchMode;
exports.validateFetchDest = validateFetchDest;
exports.validateFetchSite = validateFetchSite;
exports.validateReferer = validateReferer;
exports.setupCsrfToken = setupCsrfToken;
exports.validateSameOrigin = validateSameOrigin;
exports.validateCsrfToken = validateCsrfToken;
exports.parseHttpCookies = parseHttpCookies;
exports.extractRequestMetadata = extractRequestMetadata;
exports.extractLocales = extractLocales;
exports.negotiateResponseContent = negotiateResponseContent;
const node_crypto_1 = require("node:crypto");
const accept_1 = require("@hapi/accept");
const cookie_1 = require("cookie");
const forwarded_1 = __importDefault(require("forwarded"));
const http_errors_1 = __importDefault(require("http-errors"));
const response_js_1 = require("./response.js");
const url_js_1 = require("./url.js");
function validateHeaderValue(req, name, allowedValues) {
    const value = req.headers[name] ?? null;
    if (Array.isArray(value)) {
        throw (0, http_errors_1.default)(400, `Invalid ${name} header`);
    }
    if (!allowedValues.includes(value)) {
        throw (0, http_errors_1.default)(400, value
            ? `Forbidden ${name} header "${value}" (expected ${allowedValues})`
            : `Missing ${name} header`);
    }
}
function validateFetchMode(req, res, expectedMode) {
    validateHeaderValue(req, 'sec-fetch-mode', expectedMode);
}
function validateFetchDest(req, res, expectedDest) {
    validateHeaderValue(req, 'sec-fetch-dest', expectedDest);
}
function validateFetchSite(req, res, expectedSite) {
    validateHeaderValue(req, 'sec-fetch-site', expectedSite);
}
function validateReferer(req, res, reference, allowNull = false) {
    const referer = req.headers['referer'];
    const refererUrl = referer ? new URL(referer) : null;
    if (refererUrl ? !(0, url_js_1.urlMatch)(refererUrl, reference) : !allowNull) {
        throw (0, http_errors_1.default)(400, `Invalid referer ${referer}`);
    }
    return refererUrl;
}
async function setupCsrfToken(req, res, cookieName = 'csrf_token') {
    const csrfToken = (0, node_crypto_1.randomBytes)(8).toString('hex');
    (0, response_js_1.appendHeader)(res, 'Set-Cookie', (0, cookie_1.serialize)(cookieName, csrfToken, {
        secure: true,
        httpOnly: false,
        sameSite: 'lax',
        path: req.url?.split('?', 1)[0] || '/',
    }));
}
// CORS ensure not cross origin
function validateSameOrigin(req, res, origin, allowNull = true) {
    const reqOrigin = req.headers['origin'];
    if (reqOrigin ? reqOrigin !== origin : !allowNull) {
        throw (0, http_errors_1.default)(400, `Invalid origin ${reqOrigin}`);
    }
}
function validateCsrfToken(req, res, csrfToken, cookieName = 'csrf_token', clearCookie = false) {
    const cookies = parseHttpCookies(req);
    if (typeof csrfToken !== 'string' ||
        !csrfToken ||
        !cookies ||
        !cookieName ||
        cookies[cookieName] !== csrfToken) {
        throw (0, http_errors_1.default)(400, `Invalid CSRF token`);
    }
    if (clearCookie) {
        (0, response_js_1.appendHeader)(res, 'Set-Cookie', (0, cookie_1.serialize)(cookieName, '', {
            secure: true,
            httpOnly: false,
            sameSite: 'lax',
            maxAge: 0,
        }));
    }
}
function parseHttpCookies(req) {
    return 'cookies' in req && req.cookies // Already parsed by another middleware
        ? req.cookies
        : req.headers['cookie']
            ? (req.cookies = (0, cookie_1.parse)(req.headers['cookie']))
            : null;
}
function extractRequestMetadata(req, options) {
    const ip = extractIp(req, options);
    return {
        userAgent: req.headers['user-agent'],
        ipAddress: ip,
        port: extractPort(req, ip),
    };
}
function extractIp(req, options) {
    const trust = options?.trustProxy;
    if (trust) {
        const ips = (0, forwarded_1.default)(req);
        for (let i = 0; i < ips.length; i++) {
            const isTrusted = trust(ips[i], i);
            if (!isTrusted)
                return ips[i];
        }
        // Let's return the last ("furthest") IP address in the chain if all of them
        // are trusted. Note that this may indicate an issue with either the trust
        // function (too permissive), or the proxy configuration (one of them not
        // setting the X-Forwarded-For header).
        const ip = ips[ips.length - 1];
        if (ip)
            return ip;
    }
    // Express app compatibility (see "trust proxy" setting)
    if ('ip' in req) {
        const ip = req.ip;
        if (typeof ip === 'string')
            return ip;
    }
    const ip = req.socket.remoteAddress;
    if (ip)
        return ip;
    throw new Error('Could not determine IP address');
}
function extractPort(req, ip) {
    if (ip !== req.socket.remoteAddress) {
        // Trust the X-Forwarded-Port header only if the IP address was a trusted
        // proxied IP.
        const forwardedPort = req.headers['x-forwarded-port'];
        if (typeof forwardedPort === 'string') {
            const port = Number(forwardedPort.trim());
            if (!Number.isInteger(port) || port < 0 || port > 65535) {
                throw new Error('Invalid forwarded port');
            }
            return port;
        }
    }
    const port = req.socket.remotePort;
    if (port != null)
        return port;
    throw new Error('Could not determine port');
}
function extractLocales(req) {
    const acceptLanguage = req.headers['accept-language'];
    return acceptLanguage ? (0, accept_1.languages)(acceptLanguage) : [];
}
function negotiateResponseContent(req, types) {
    const type = (0, accept_1.mediaType)(req.headers['accept'], types);
    if (type)
        return type;
    return undefined;
}
