// @atproto/oauth-provider/dist/lib/http/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendHeader = appendHeader;
exports.writeRedirect = writeRedirect;
exports.writeStream = writeStream;
exports.writeBuffer = writeBuffer;
exports.toJsonBuffer = toJsonBuffer;
exports.writeJson = writeJson;
exports.staticJsonMiddleware = staticJsonMiddleware;
exports.writeHtml = writeHtml;
exports.cacheControlMiddleware = cacheControlMiddleware;
const node_stream_1 = require("node:stream");
const security_headers_js_1 = require("./security-headers.js");
function appendHeader(res, header, value) {
    const existing = res.getHeader(header);
    if (existing == null) {
        res.setHeader(header, value);
    }
    else {
        const arr = Array.isArray(existing) ? existing : [String(existing)];
        res.setHeader(header, arr.concat(value));
    }
}
function writeRedirect(res, url, status = 302) {
    res.writeHead(status, { Location: url }).end();
}
function writeStream(res, stream, { status = 200, contentType = 'application/octet-stream', } = {}) {
    res.statusCode = status;
    res.setHeader('content-type', contentType);
    if (res.req.method === 'HEAD') {
        res.end();
        stream.destroy();
    }
    else {
        (0, node_stream_1.pipeline)([stream, res], (_err) => {
            // The error will be propagated through the streams
        });
    }
}
function writeBuffer(res, chunk, opts) {
    if (opts?.status != null)
        res.statusCode = opts.status;
    res.setHeader('content-type', opts?.contentType || 'application/octet-stream');
    res.end(chunk);
}
function toJsonBuffer(value) {
    try {
        return Buffer.from(JSON.stringify(value));
    }
    catch (cause) {
        throw new Error(`Failed to serialize as JSON`, { cause });
    }
}
function writeJson(res, payload, { contentType = 'application/json', ...options } = {}) {
    const buffer = toJsonBuffer(payload);
    writeBuffer(res, buffer, { ...options, contentType });
}
function staticJsonMiddleware(value, { contentType = 'application/json', ...options } = {}) {
    const buffer = toJsonBuffer(value);
    const staticOptions = { ...options, contentType };
    return function (req, res) {
        writeBuffer(res, buffer, staticOptions);
    };
}
function writeHtml(res, html, { contentType = 'text/html', ...options } = {}) {
    // HTML pages should always be served with safety protection headers
    (0, security_headers_js_1.setSecurityHeaders)(res, options);
    writeBuffer(res, html, { ...options, contentType });
}
function cacheControlMiddleware(maxAge) {
    const header = `max-age=${maxAge}`;
    return function (req, res, next) {
        res.setHeader('Cache-Control', header);
        next();
    };
}
