// @atproto/oauth-provider/dist/lib/http/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const context_js_1 = require("./context.js");
const middleware_js_1 = require("./middleware.js");
const route_js_1 = require("./route.js");
class Router {
    url;
    middlewares = [];
    constructor(url) {
        this.url = url;
    }
    use(...middlewares) {
        this.middlewares.push(...middlewares);
        return this;
    }
    all(path, ...mw) {
        return this.addRoute('*', path, ...mw);
    }
    get(path, ...mw) {
        return this.addRoute('GET', path, ...mw);
    }
    post(path, ...mw) {
        return this.addRoute('POST', path, ...mw);
    }
    options(path, ...mw) {
        return this.addRoute('OPTIONS', path, ...mw);
    }
    addRoute(method, path, ...mw) {
        return this.use((0, route_js_1.createRoute)(method, path, ...mw));
    }
    /**
     * @returns router middleware which dispatches a route matching the request.
     */
    buildHandler() {
        const routerUrl = this.url;
        // Calling next('router') from a middleware will skip all the remaining
        // middlewares in the stack.
        const middleware = (0, middleware_js_1.combineMiddlewares)(this.middlewares, {
            skipKeyword: 'router',
        });
        return (0, middleware_js_1.asHandler)(function (req, res, next) {
            // Make sure that the context contains a "url". This will allow the add()
            // method to match routes based on the pathname and will allow routes to
            // access the query params (through this.url.searchParams).
            let url;
            if (!routerUrl &&
                this != null &&
                typeof this === 'object' &&
                'url' in this &&
                this.url instanceof URL) {
                // If the context already contains a "url" (router inside router), let's
                // use it.
                url = this.url;
            }
            else {
                // Parse the URL using node's URL parser.
                try {
                    const protocol = routerUrl?.protocol || 'https:';
                    const host = req.headers.host || routerUrl?.host || 'localhost';
                    const pathname = req.url || '/';
                    url = new URL(pathname, `${protocol}//${host}`);
                }
                catch (cause) {
                    const error = cause instanceof Error ? cause : new Error('Invalid URL', { cause });
                    return next(Object.assign(error, { status: 400, statusCode: 400 }));
                }
            }
            const context = (0, context_js_1.subCtx)(this, 'url', url);
            middleware.call(context, req, res, next);
        });
    }
}
exports.Router = Router;
