// @atproto/oauth-provider/dist/lib/http/security-headers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossOriginOpenerPolicy = exports.CrossOriginResourcePolicy = exports.CrossOriginEmbedderPolicy = void 0;
exports.setSecurityHeaders = setSecurityHeaders;
const index_js_1 = require("../csp/index.js");
/**
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy COEP on MDN}
 */
var CrossOriginEmbedderPolicy;
(function (CrossOriginEmbedderPolicy) {
    CrossOriginEmbedderPolicy["unsafeNone"] = "unsafe-none";
    CrossOriginEmbedderPolicy["requireCorp"] = "require-corp";
    CrossOriginEmbedderPolicy["credentialless"] = "credentialless";
})(CrossOriginEmbedderPolicy || (exports.CrossOriginEmbedderPolicy = CrossOriginEmbedderPolicy = {}));
/**
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Resource-Policy CORP on MDN}
 */
var CrossOriginResourcePolicy;
(function (CrossOriginResourcePolicy) {
    CrossOriginResourcePolicy["sameSite"] = "same-site";
    CrossOriginResourcePolicy["sameOrigin"] = "same-origin";
    CrossOriginResourcePolicy["crossOrigin"] = "cross-origin";
})(CrossOriginResourcePolicy || (exports.CrossOriginResourcePolicy = CrossOriginResourcePolicy = {}));
/**
 * @see {@link https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Opener-Policy COOP on MDN}
 */
var CrossOriginOpenerPolicy;
(function (CrossOriginOpenerPolicy) {
    CrossOriginOpenerPolicy["unsafeNone"] = "unsafe-none";
    CrossOriginOpenerPolicy["sameOriginAllowPopups"] = "same-origin-allow-popups";
    CrossOriginOpenerPolicy["sameOrigin"] = "same-origin";
    CrossOriginOpenerPolicy["noopenerAllowPopups"] = "noopener-allow-popups";
})(CrossOriginOpenerPolicy || (exports.CrossOriginOpenerPolicy = CrossOriginOpenerPolicy = {}));
function setSecurityHeaders(res, { csp = { 'default-src': ["'none'"] }, coep = CrossOriginEmbedderPolicy.requireCorp, corp = CrossOriginResourcePolicy.sameOrigin, coop = CrossOriginOpenerPolicy.sameOrigin, hsts = { maxAge: 63072000 }, }) {
    // @NOTE Never set CSP through http-equiv meta as not all directives will
    // be honored. Always set it through the Content-Security-Policy header.
    const cspString = (0, index_js_1.buildCsp)(csp);
    if (cspString) {
        res.setHeader('Content-Security-Policy', cspString);
    }
    res.setHeader('Cross-Origin-Embedder-Policy', coep);
    res.setHeader('Cross-Origin-Resource-Policy', corp);
    res.setHeader('Cross-Origin-Opener-Policy', coop);
    if (hsts) {
        res.setHeader('Strict-Transport-Security', buildHstsValue(hsts));
    }
    // @TODO: make these headers configurable (?)
    res.setHeader('Permissions-Policy', 'otp-credentials=*, document-domain=()');
    res.setHeader('Referrer-Policy', 'same-origin');
    res.setHeader('X-Frame-Options', 'DENY');
    res.setHeader('X-Content-Type-Options', 'nosniff');
    res.setHeader('X-XSS-Protection', '0');
}
function buildHstsValue(config) {
    let value = `max-age=${config.maxAge}`;
    if (config.includeSubDomains)
        value += '; includeSubDomains';
    if (config.preload)
        value += '; preload';
    return value;
}
