// @atproto/oauth-provider/dist/dpop/dpop-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DpopManager = exports.dpopManagerOptionsSchema = exports.DpopNonce = void 0;
const node_crypto_1 = require("node:crypto");
const jose_1 = require("jose");
const zod_1 = require("zod");
const constants_js_1 = require("../constants.js");
const invalid_dpop_proof_error_js_1 = require("../errors/invalid-dpop-proof-error.js");
const use_dpop_nonce_error_js_1 = require("../errors/use-dpop-nonce-error.js");
const dpop_nonce_js_1 = require("./dpop-nonce.js");
Object.defineProperty(exports, "DpopNonce", { enumerable: true, get: function () { return dpop_nonce_js_1.DpopNonce; } });
const { JOSEError } = jose_1.errors;
exports.dpopManagerOptionsSchema = zod_1.z.object({
    /**
     * Set this to `false` to disable the use of nonces in DPoP proofs. Set this
     * to a secret Uint8Array or hex encoded string to use a predictable seed for
     * all nonces (typically useful when multiple instances are running). Leave
     * undefined to generate a random seed at startup.
     */
    dpopSecret: zod_1.z.union([zod_1.z.literal(false), dpop_nonce_js_1.dpopSecretSchema]).optional(),
    dpopRotationInterval: dpop_nonce_js_1.rotationIntervalSchema.optional(),
});
class DpopManager {
    dpopNonce;
    constructor(options = {}) {
        const { dpopSecret, dpopRotationInterval } = exports.dpopManagerOptionsSchema.parse(options);
        this.dpopNonce =
            dpopSecret === false
                ? undefined
                : new dpop_nonce_js_1.DpopNonce(dpopSecret, dpopRotationInterval);
    }
    nextNonce() {
        return this.dpopNonce?.next();
    }
    /**
     * @see {@link https://datatracker.ietf.org/doc/html/rfc9449#section-4.3}
     */
    async checkProof(proof, htm, // HTTP Method
    htu, // HTTP URL
    accessToken) {
        if (Array.isArray(proof) && proof.length === 1) {
            proof = proof[0];
        }
        if (!proof || typeof proof !== 'string') {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP proof required');
        }
        const { protectedHeader, payload } = await (0, jose_1.jwtVerify)(proof, jose_1.EmbeddedJWK, {
            typ: 'dpop+jwt',
            maxTokenAge: 10,
            clockTolerance: constants_js_1.DPOP_NONCE_MAX_AGE / 1e3,
            requiredClaims: ['iat', 'jti'],
        }).catch((err) => {
            const message = err instanceof JOSEError
                ? `Invalid DPoP proof (${err.message})`
                : 'Invalid DPoP proof';
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError(message, err);
        });
        if (!payload.jti || typeof payload.jti !== 'string') {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('Invalid or missing jti property');
        }
        // Note rfc9110#section-9.1 states that the method name is case-sensitive
        if (!htm || htm !== payload['htm']) {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP htm mismatch');
        }
        if (payload['nonce'] !== undefined &&
            typeof payload['nonce'] !== 'string') {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP nonce must be a string');
        }
        if (!payload['nonce'] && this.dpopNonce) {
            throw new use_dpop_nonce_error_js_1.UseDpopNonceError();
        }
        if (payload['nonce'] && !this.dpopNonce?.check(payload['nonce'])) {
            throw new use_dpop_nonce_error_js_1.UseDpopNonceError('DPoP nonce mismatch');
        }
        const htuNorm = normalizeHtu(htu);
        if (!htuNorm) {
            throw new TypeError('Invalid "htu" argument');
        }
        if (htuNorm !== normalizeHtu(payload['htu'])) {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP htu mismatch');
        }
        if (accessToken) {
            const athBuffer = (0, node_crypto_1.createHash)('sha256').update(accessToken).digest();
            if (payload['ath'] !== athBuffer.toString('base64url')) {
                throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP ath mismatch');
            }
        }
        else if (payload['ath']) {
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError('DPoP ath not allowed');
        }
        try {
            return {
                protectedHeader,
                payload,
                jkt: await (0, jose_1.calculateJwkThumbprint)(protectedHeader['jwk'], 'sha256'), // EmbeddedJWK
            };
        }
        catch (err) {
            const message = err instanceof JOSEError ? err.message : 'Failed to calculate jkt';
            throw new invalid_dpop_proof_error_js_1.InvalidDpopProofError(message, err);
        }
    }
}
exports.DpopManager = DpopManager;
/**
 * @note
 * > The htu claim matches the HTTP URI value for the HTTP request in which the
 * > JWT was received, ignoring any query and fragment parts.
 *
 * > To reduce the likelihood of false negatives, servers SHOULD employ
 * > syntax-based normalization (Section 6.2.2 of [RFC3986]) and scheme-based
 * > normalization (Section 6.2.3 of [RFC3986]) before comparing the htu claim.
 * @see {@link https://datatracker.ietf.org/doc/html/rfc9449#section-4.3 | RFC9449 section 4.3. Checking DPoP Proofs}
 */
function normalizeHtu(htu) {
    // Optimization
    if (!htu)
        return null;
    try {
        const url = new URL(String(htu));
        url.hash = '';
        url.search = '';
        return url.href;
    }
    catch {
        return null;
    }
}
