// @atproto/oauth-provider/dist/account/account-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountManager = void 0;
const oauth_types_1 = require("@atproto/oauth-types");
const invalid_request_error_js_1 = require("../errors/invalid-request-error.js");
const hcaptcha_js_1 = require("../lib/hcaptcha.js");
const function_js_1 = require("../lib/util/function.js");
const time_js_1 = require("../lib/util/time.js");
const TIMING_ATTACK_MITIGATION_DELAY = 400;
const BRUTE_FORCE_MITIGATION_DELAY = 300;
class AccountManager {
    store;
    hooks;
    inviteCodeRequired;
    hcaptchaClient;
    constructor(issuer, store, hooks, customization) {
        this.store = store;
        this.hooks = hooks;
        this.inviteCodeRequired = customization.inviteCodeRequired !== false;
        this.hcaptchaClient = customization.hcaptcha
            ? new hcaptcha_js_1.HCaptchaClient(new URL(issuer).hostname, customization.hcaptcha)
            : undefined;
    }
    async processHcaptchaToken(input, deviceId, deviceMetadata) {
        if (!this.hcaptchaClient) {
            return undefined;
        }
        if (!input.hcaptchaToken) {
            throw new invalid_request_error_js_1.InvalidRequestError('hCaptcha token is required');
        }
        const tokens = this.hcaptchaClient.buildClientTokens(deviceMetadata.ipAddress, input.handle, deviceMetadata.userAgent);
        const result = await this.hcaptchaClient
            .verify('signup', input.hcaptchaToken, deviceMetadata.ipAddress, tokens)
            .catch((err) => {
            throw invalid_request_error_js_1.InvalidRequestError.from(err, 'hCaptcha verification failed');
        });
        await (0, function_js_1.callAsync)(this.hooks.onHcaptchaResult, {
            input,
            deviceId,
            deviceMetadata,
            tokens,
            result,
        });
        try {
            this.hcaptchaClient.checkVerifyResult(result, tokens);
        }
        catch (err) {
            throw invalid_request_error_js_1.InvalidRequestError.from(err, 'hCaptcha verification failed');
        }
        return result;
    }
    async enforceInviteCode(input, _deviceId, _deviceMetadata) {
        if (!this.inviteCodeRequired) {
            return undefined;
        }
        if (!input.inviteCode) {
            throw new invalid_request_error_js_1.InvalidRequestError('Invite code is required');
        }
        return input.inviteCode;
    }
    async buildSignupData(input, deviceId, deviceMetadata) {
        const [hcaptchaResult, inviteCode] = await Promise.all([
            this.processHcaptchaToken(input, deviceId, deviceMetadata),
            this.enforceInviteCode(input, deviceId, deviceMetadata),
        ]);
        return { ...input, hcaptchaResult, inviteCode };
    }
    async signUp(input, deviceId, deviceMetadata) {
        await (0, function_js_1.callAsync)(this.hooks.onSignupAttempt, {
            input,
            deviceId,
            deviceMetadata,
        });
        const data = await this.buildSignupData(input, deviceId, deviceMetadata);
        // Mitigation against brute forcing email of users.
        // @TODO Add rate limit to all the OAuth routes.
        return (0, time_js_1.constantTime)(BRUTE_FORCE_MITIGATION_DELAY, async () => {
            let account;
            try {
                account = await this.store.createAccount(data);
            }
            catch (err) {
                throw invalid_request_error_js_1.InvalidRequestError.from(err, 'Account creation failed');
            }
            try {
                const info = await this.store.addDeviceAccount(deviceId, account.sub, false);
                await (0, function_js_1.callAsync)(this.hooks.onSignedUp, {
                    data,
                    info,
                    account,
                    deviceId,
                    deviceMetadata,
                });
                return { account, info };
            }
            catch (err) {
                throw invalid_request_error_js_1.InvalidRequestError.from(err, 'Something went wrong, try singing-in');
            }
        });
    }
    async signIn(data, deviceId, deviceMetadata) {
        return (0, time_js_1.constantTime)(TIMING_ATTACK_MITIGATION_DELAY, async () => {
            try {
                const account = await this.store.authenticateAccount(data);
                const info = await this.store.addDeviceAccount(deviceId, account.sub, data.remember);
                await (0, function_js_1.callAsync)(this.hooks.onSignedIn, {
                    data,
                    info,
                    account,
                    deviceId,
                    deviceMetadata,
                });
                return { account, info };
            }
            catch (err) {
                throw invalid_request_error_js_1.InvalidRequestError.from(err, 'Unable to sign-in due to an unexpected server error');
            }
        });
    }
    async get(deviceId, sub) {
        const result = await this.store.getDeviceAccount(deviceId, sub);
        if (result)
            return result;
        throw new invalid_request_error_js_1.InvalidRequestError(`Account not found`);
    }
    async addAuthorizedClient(deviceId, account, client, _clientAuth) {
        // "Loopback" clients are not distinguishable from one another.
        if ((0, oauth_types_1.isOAuthClientIdLoopback)(client.id))
            return;
        await this.store.addAuthorizedClient(deviceId, account.sub, client.id);
    }
    async list(deviceId) {
        const results = await this.store.listDeviceAccounts(deviceId);
        return results.filter((result) => result.info.remembered);
    }
    async resetPasswordRequest(data) {
        return (0, time_js_1.constantTime)(TIMING_ATTACK_MITIGATION_DELAY, async () => {
            await this.store.resetPasswordRequest(data);
        });
    }
    async resetPasswordConfirm(data) {
        return (0, time_js_1.constantTime)(TIMING_ATTACK_MITIGATION_DELAY, async () => {
            await this.store.resetPasswordConfirm(data);
        });
    }
    async verifyHandleAvailability(handle) {
        return (0, time_js_1.constantTime)(TIMING_ATTACK_MITIGATION_DELAY, async () => {
            return this.store.verifyHandleAvailability(handle);
        });
    }
}
exports.AccountManager = AccountManager;
