// @atproto/oauth-provider/dist/account/account-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAccountStore = exports.SecondAuthenticationFactorRequiredError = exports.InvalidRequestError = exports.HandleUnavailableError = exports.resetPasswordConfirmDataSchema = exports.resetPasswordRequestDataSchema = exports.createAccountDataSchema = exports.authenticateAccountDataSchema = exports.inviteCodeSchema = exports.emailSchema = exports.handleSchema = exports.tokenSchema = exports.newPasswordSchema = exports.oldPasswordSchema = void 0;
exports.asAccountStore = asAccountStore;
const address_1 = require("@hapi/address");
const disposable_email_domains_js_1 = require("disposable-email-domains-js");
const zod_1 = require("zod");
const syntax_1 = require("@atproto/syntax");
const locale_js_1 = require("../lib/locale.js");
const type_js_1 = require("../lib/util/type.js");
const oauth_errors_js_1 = require("../oauth-errors.js");
Object.defineProperty(exports, "HandleUnavailableError", { enumerable: true, get: function () { return oauth_errors_js_1.HandleUnavailableError; } });
Object.defineProperty(exports, "InvalidRequestError", { enumerable: true, get: function () { return oauth_errors_js_1.InvalidRequestError; } });
Object.defineProperty(exports, "SecondAuthenticationFactorRequiredError", { enumerable: true, get: function () { return oauth_errors_js_1.SecondAuthenticationFactorRequiredError; } });
// @NOTE Change the length here to force stronger passwords (through a reset)
exports.oldPasswordSchema = zod_1.z.string().min(1);
exports.newPasswordSchema = zod_1.z.string().min(8);
exports.tokenSchema = zod_1.z
    .string()
    .regex(/^[A-Z2-7]{5}-[A-Z2-7]{5}$/, 'Invalid token format');
exports.handleSchema = zod_1.z
    .string()
    // @NOTE: We only check against validity towards ATProto's syntax. Additional
    // rules may be imposed by the store implementation.
    .superRefine((value, ctx) => {
    try {
        (0, syntax_1.ensureValidHandle)(value);
    }
    catch (err) {
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: err instanceof Error ? err.message : 'Invalid handle',
        });
    }
})
    .transform(syntax_1.normalizeHandle);
exports.emailSchema = zod_1.z
    .string()
    .email()
    // @NOTE using @hapi/address here, in addition to the email() check to ensure
    // compatibility with the current email validation in the PDS's account
    // manager
    .refine(address_1.isEmailValid, {
    message: 'Invalid email address',
})
    .refine((email) => !(0, disposable_email_domains_js_1.isDisposableEmail)(email), {
    message: 'Disposable email addresses are not allowed',
})
    .transform((value) => value.toLowerCase());
exports.inviteCodeSchema = zod_1.z.string().min(1);
exports.authenticateAccountDataSchema = zod_1.z
    .object({
    locale: locale_js_1.localeSchema,
    username: zod_1.z.string(),
    password: exports.oldPasswordSchema,
    emailOtp: exports.tokenSchema.optional(),
})
    .strict();
exports.createAccountDataSchema = zod_1.z
    .object({
    locale: locale_js_1.localeSchema,
    handle: exports.handleSchema,
    email: exports.emailSchema,
    password: zod_1.z.intersection(exports.oldPasswordSchema, exports.newPasswordSchema),
    inviteCode: exports.inviteCodeSchema.optional(),
})
    .strict();
exports.resetPasswordRequestDataSchema = zod_1.z
    .object({
    locale: locale_js_1.localeSchema,
    email: exports.emailSchema,
})
    .strict();
exports.resetPasswordConfirmDataSchema = zod_1.z
    .object({
    token: exports.tokenSchema,
    password: zod_1.z.intersection(exports.oldPasswordSchema, exports.newPasswordSchema),
})
    .strict();
exports.isAccountStore = (0, type_js_1.buildInterfaceChecker)([
    'createAccount',
    'authenticateAccount',
    'addAuthorizedClient',
    'addDeviceAccount',
    'getDeviceAccount',
    'removeDeviceAccount',
    'listDeviceAccounts',
    'resetPasswordRequest',
    'resetPasswordConfirm',
    'verifyHandleAvailability',
]);
function asAccountStore(implementation) {
    if (!implementation || !(0, exports.isAccountStore)(implementation)) {
        throw new Error('Invalid AccountStore implementation');
    }
    return implementation;
}
