// @atproto/oauth-provider/dist/lib/html/build-document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildDocument = exports.isLinkRel = void 0;
const html_js_1 = require("./html.js");
const tags_js_1 = require("./tags.js");
/**
 * @see {@link https://developer.mozilla.org/fr/docs/Web/HTML/Attributes/rel}
 */
const ALLOWED_LINK_REL_VALUES = Object.freeze([
    'alternate',
    'author',
    'canonical',
    'dns-prefetch',
    'external',
    'expect',
    'help',
    'icon',
    'license',
    'manifest',
    'me',
    'modulepreload',
    'next',
    'pingback',
    'preconnect',
    'prefetch',
    'preload',
    'prerender',
    'prev',
    'privacy-policy',
    'search',
    'stylesheet',
    'terms-of-service',
]);
const isLinkRel = (rel) => ALLOWED_LINK_REL_VALUES.includes(rel);
exports.isLinkRel = isLinkRel;
const defaultViewport = (0, tags_js_1.html) `<meta
  name="viewport"
  content="width=device-width, initial-scale=1.0"
/>`;
const buildDocument = ({ htmlAttrs, head, title, body, bodyAttrs, base, meta, links, preloads, scripts, styles, }) => (0, tags_js_1.html) `<!doctype html>
<html${attrsToHtml(htmlAttrs)}>
  <head>
    <meta charset="UTF-8" />
    ${title && (0, tags_js_1.html) `<title>${title}</title>`}
    ${base && (0, tags_js_1.html) `<base href="${base.href}" />`}
    ${meta?.some(isViewportMeta) ? null : defaultViewport}
    ${meta?.map(metaToHtml)}
    ${preloads?.map(linkPreload)}
    ${links?.map(linkToHtml)}
    ${head}
    ${styles?.map(styleToHtml)}
  </head>
  <body${attrsToHtml(bodyAttrs)}>${body}${scripts?.map(scriptToHtml)}</body>
</html>`;
exports.buildDocument = buildDocument;
function isViewportMeta(attrs) {
    return 'name' in attrs && attrs.name === 'viewport';
}
function linkToHtml(attrs) {
    return (0, tags_js_1.html) `<link${attrsToHtml(attrs)} />`;
}
function metaToHtml(attrs) {
    return (0, tags_js_1.html) `<meta${attrsToHtml(attrs)} />`;
}
function* attrsToHtml(attrs) {
    if (attrs) {
        for (const [name, value] of Object.entries(attrs)) {
            if (value == null)
                continue;
            else if (value === false)
                continue;
            else if (value === true)
                yield (0, tags_js_1.html) ` ${name}`;
            else
                yield (0, tags_js_1.html) ` ${name}="${value}"`;
        }
    }
}
function linkPreload(asset) {
    const [path] = asset.url.split('?', 2);
    if (path.endsWith('.js')) {
        return (0, tags_js_1.html) `<link rel="modulepreload" href="${asset.url}" />`;
    }
    if (path.endsWith('.css')) {
        return (0, tags_js_1.html) `<link rel="preload" href="${asset.url}" as="style" />`;
    }
    return undefined;
}
function scriptToHtml(script) {
    return script instanceof html_js_1.Html
        ? // prettier-ignore
            (0, tags_js_1.html) `<script>${script}</script>` // hash validity requires no space around the content
        : (0, tags_js_1.html) `<script type="module" src="${script.url}"></script>`;
}
function styleToHtml(style) {
    return style instanceof html_js_1.Html
        ? // prettier-ignore
            (0, tags_js_1.html) `<style>${style}</style>` // hash validity requires no space around the content
        : (0, tags_js_1.html) `<link rel="stylesheet" href="${style.url}" />`;
}
