// @atproto/xrpc-server/dist/stream/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloseCode = exports.DisconnectError = exports.frameHeader = exports.errorFrameBody = exports.errorFrameHeader = exports.messageFrameHeader = exports.FrameType = void 0;
const zod_1 = require("zod");
var FrameType;
(function (FrameType) {
    FrameType[FrameType["Message"] = 1] = "Message";
    FrameType[FrameType["Error"] = -1] = "Error";
})(FrameType || (exports.FrameType = FrameType = {}));
exports.messageFrameHeader = zod_1.z.object({
    op: zod_1.z.literal(FrameType.Message), // Frame op
    t: zod_1.z.string().optional(), // Message body type discriminator
});
exports.errorFrameHeader = zod_1.z.object({
    op: zod_1.z.literal(FrameType.Error),
});
exports.errorFrameBody = zod_1.z.object({
    error: zod_1.z.string(), // Error code
    message: zod_1.z.string().optional(), // Error message
});
exports.frameHeader = zod_1.z.union([exports.messageFrameHeader, exports.errorFrameHeader]);
class DisconnectError extends Error {
    constructor(wsCode = CloseCode.Policy, xrpcCode) {
        super();
        Object.defineProperty(this, "wsCode", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: wsCode
        });
        Object.defineProperty(this, "xrpcCode", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: xrpcCode
        });
    }
}
exports.DisconnectError = DisconnectError;
// https://www.rfc-editor.org/rfc/rfc6455#section-7.4.1
var CloseCode;
(function (CloseCode) {
    CloseCode[CloseCode["Normal"] = 1000] = "Normal";
    CloseCode[CloseCode["Abnormal"] = 1006] = "Abnormal";
    CloseCode[CloseCode["Policy"] = 1008] = "Policy";
})(CloseCode || (exports.CloseCode = CloseCode = {}));
