// @atproto/xrpc-server/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodNotImplementedError = exports.UpstreamTimeoutError = exports.NotEnoughResourcesError = exports.UpstreamFailureError = exports.InternalServerError = exports.RateLimitExceededError = exports.ForbiddenError = exports.AuthRequiredError = exports.InvalidRequestError = exports.XRPCError = exports.ResponseType = exports.isShared = exports.handlerError = exports.handlerPipeThrough = exports.handlerPipeThroughStream = exports.handlerPipeThroughBuffer = exports.handlerSuccess = exports.headersSchema = exports.handlerAuth = exports.handlerInput = void 0;
exports.isHandlerError = isHandlerError;
exports.isHandlerPipeThroughBuffer = isHandlerPipeThroughBuffer;
exports.isHandlerPipeThroughStream = isHandlerPipeThroughStream;
const node_stream_1 = require("node:stream");
const http_errors_1 = require("http-errors");
const zod_1 = require("zod");
const xrpc_1 = require("@atproto/xrpc");
Object.defineProperty(exports, "ResponseType", { enumerable: true, get: function () { return xrpc_1.ResponseType; } });
exports.handlerInput = zod_1.z.object({
    encoding: zod_1.z.string(),
    body: zod_1.z.any(),
});
exports.handlerAuth = zod_1.z.object({
    credentials: zod_1.z.any(),
    artifacts: zod_1.z.any(),
});
exports.headersSchema = zod_1.z.record(zod_1.z.string());
exports.handlerSuccess = zod_1.z.object({
    encoding: zod_1.z.string(),
    body: zod_1.z.any(),
    headers: exports.headersSchema.optional(),
});
exports.handlerPipeThroughBuffer = zod_1.z.object({
    encoding: zod_1.z.string(),
    buffer: zod_1.z.instanceof(Buffer),
    headers: exports.headersSchema.optional(),
});
exports.handlerPipeThroughStream = zod_1.z.object({
    encoding: zod_1.z.string(),
    stream: zod_1.z.instanceof(node_stream_1.Readable),
    headers: exports.headersSchema.optional(),
});
exports.handlerPipeThrough = zod_1.z.union([
    exports.handlerPipeThroughBuffer,
    exports.handlerPipeThroughStream,
]);
exports.handlerError = zod_1.z.object({
    status: zod_1.z.number(),
    error: zod_1.z.string().optional(),
    message: zod_1.z.string().optional(),
});
const isShared = (opts) => {
    return typeof opts['name'] === 'string';
};
exports.isShared = isShared;
class XRPCError extends Error {
    constructor(type, errorMessage, customErrorName, options) {
        super(errorMessage, options);
        Object.defineProperty(this, "type", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: type
        });
        Object.defineProperty(this, "errorMessage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: errorMessage
        });
        Object.defineProperty(this, "customErrorName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: customErrorName
        });
    }
    get payload() {
        return {
            error: this.customErrorName ?? this.typeName,
            message: this.type === xrpc_1.ResponseType.InternalServerError
                ? this.typeStr // Do not respond with error details for 500s
                : this.errorMessage || this.typeStr,
        };
    }
    get typeName() {
        return xrpc_1.ResponseTypeNames[this.type];
    }
    get typeStr() {
        return xrpc_1.ResponseTypeStrings[this.type];
    }
    static fromError(cause) {
        if (cause instanceof XRPCError) {
            return cause;
        }
        if (cause instanceof xrpc_1.XRPCError) {
            return new XRPCError(cause.status, cause.message, cause.error, { cause });
        }
        if ((0, http_errors_1.isHttpError)(cause)) {
            return new XRPCError(cause.status, cause.message, cause.name, { cause });
        }
        if (isHandlerError(cause)) {
            return this.fromHandlerError(cause);
        }
        if (cause instanceof Error) {
            return new InternalServerError(cause.message, undefined, { cause });
        }
        return new InternalServerError('Unexpected internal server error', undefined, { cause });
    }
    static fromHandlerError(err) {
        return new XRPCError(err.status, err.message, err.error, { cause: err });
    }
}
exports.XRPCError = XRPCError;
function isHandlerError(v) {
    return (!!v &&
        typeof v === 'object' &&
        typeof v['status'] === 'number' &&
        (v['error'] === undefined || typeof v['error'] === 'string') &&
        (v['message'] === undefined || typeof v['message'] === 'string'));
}
function isHandlerPipeThroughBuffer(v) {
    // We only need to discriminate between possible HandlerOutput values
    return v['buffer'] !== undefined;
}
function isHandlerPipeThroughStream(v) {
    // We only need to discriminate between possible HandlerOutput values
    return v['stream'] !== undefined;
}
class InvalidRequestError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.InvalidRequest, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.InvalidRequest);
    }
}
exports.InvalidRequestError = InvalidRequestError;
class AuthRequiredError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.AuthRequired, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.AuthRequired);
    }
}
exports.AuthRequiredError = AuthRequiredError;
class ForbiddenError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.Forbidden, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError && instance.type === xrpc_1.ResponseType.Forbidden);
    }
}
exports.ForbiddenError = ForbiddenError;
class RateLimitExceededError extends XRPCError {
    constructor(status, errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.RateLimitExceeded, errorMessage, customErrorName, options);
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: status
        });
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.RateLimitExceeded);
    }
}
exports.RateLimitExceededError = RateLimitExceededError;
class InternalServerError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.InternalServerError, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.InternalServerError);
    }
}
exports.InternalServerError = InternalServerError;
class UpstreamFailureError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.UpstreamFailure, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.UpstreamFailure);
    }
}
exports.UpstreamFailureError = UpstreamFailureError;
class NotEnoughResourcesError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.NotEnoughResources, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.NotEnoughResources);
    }
}
exports.NotEnoughResourcesError = NotEnoughResourcesError;
class UpstreamTimeoutError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.UpstreamTimeout, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.UpstreamTimeout);
    }
}
exports.UpstreamTimeoutError = UpstreamTimeoutError;
class MethodNotImplementedError extends XRPCError {
    constructor(errorMessage, customErrorName, options) {
        super(xrpc_1.ResponseType.MethodNotImplemented, errorMessage, customErrorName, options);
    }
    [Symbol.hasInstance](instance) {
        return (instance instanceof XRPCError &&
            instance.type === xrpc_1.ResponseType.MethodNotImplemented);
    }
}
exports.MethodNotImplementedError = MethodNotImplementedError;
