// @atproto/xrpc-server/dist/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUrlNsid = exports.parseReqNsid = exports.ServerTimer = void 0;
exports.decodeQueryParams = decodeQueryParams;
exports.decodeQueryParam = decodeQueryParam;
exports.getQueryParams = getQueryParams;
exports.validateInput = validateInput;
exports.validateOutput = validateOutput;
exports.normalizeMime = normalizeMime;
exports.processBodyAsBytes = processBodyAsBytes;
exports.serverTimingHeader = serverTimingHeader;
const node_assert_1 = __importDefault(require("node:assert"));
const node_stream_1 = require("node:stream");
const mime_types_1 = __importDefault(require("mime-types"));
const common_1 = require("@atproto/common");
const lexicon_1 = require("@atproto/lexicon");
const xrpc_1 = require("@atproto/xrpc");
const types_1 = require("./types");
function decodeQueryParams(def, params) {
    const decoded = {};
    for (const k in params) {
        const val = params[k];
        const property = def.parameters?.properties?.[k];
        if (property) {
            if (property.type === 'array') {
                const vals = [];
                decoded[k] = val
                    ? vals
                        .concat(val) // Cast to array
                        .flatMap((v) => decodeQueryParam(property.items.type, v) ?? [])
                    : undefined;
            }
            else {
                decoded[k] = decodeQueryParam(property.type, val);
            }
        }
    }
    return decoded;
}
function decodeQueryParam(type, value) {
    if (!value) {
        return undefined;
    }
    if (type === 'string' || type === 'datetime') {
        return String(value);
    }
    if (type === 'float') {
        return Number(String(value));
    }
    else if (type === 'integer') {
        return parseInt(String(value), 10) || 0;
    }
    else if (type === 'boolean') {
        return value === 'true';
    }
}
function getQueryParams(url = '') {
    const { searchParams } = new URL(url ?? '', 'http://x');
    const result = {};
    for (const key of searchParams.keys()) {
        result[key] = searchParams.getAll(key);
        if (result[key].length === 1) {
            result[key] = result[key][0];
        }
    }
    return result;
}
function validateInput(nsid, def, req, opts, lexicons) {
    // request expectation
    const bodyPresence = getBodyPresence(req);
    if (bodyPresence === 'present' && (def.type !== 'procedure' || !def.input)) {
        throw new types_1.InvalidRequestError(`A request body was provided when none was expected`);
    }
    if (def.type === 'query') {
        return;
    }
    if (bodyPresence === 'missing' && def.input) {
        throw new types_1.InvalidRequestError(`A request body is expected but none was provided`);
    }
    // mimetype
    const inputEncoding = normalizeMime(req.headers['content-type'] || '');
    if (def.input?.encoding &&
        (!inputEncoding || !isValidEncoding(def.input?.encoding, inputEncoding))) {
        if (!inputEncoding) {
            throw new types_1.InvalidRequestError(`Request encoding (Content-Type) required but not provided`);
        }
        else {
            throw new types_1.InvalidRequestError(`Wrong request encoding (Content-Type): ${inputEncoding}`);
        }
    }
    if (!inputEncoding) {
        // no input body
        return undefined;
    }
    // if input schema, validate
    if (def.input?.schema) {
        try {
            const lexBody = req.body ? (0, lexicon_1.jsonToLex)(req.body) : req.body;
            req.body = lexicons.assertValidXrpcInput(nsid, lexBody);
        }
        catch (e) {
            throw new types_1.InvalidRequestError(e instanceof Error ? e.message : String(e));
        }
    }
    // if middleware already got the body, we pass that along as input
    // otherwise, we pass along a decoded readable stream
    let body;
    if (req.readableEnded) {
        body = req.body;
    }
    else {
        body = decodeBodyStream(req, opts.blobLimit);
    }
    return {
        encoding: inputEncoding,
        body,
    };
}
function validateOutput(nsid, def, output, lexicons) {
    // initial validation
    if (output) {
        types_1.handlerSuccess.parse(output);
    }
    // response expectation
    if (output?.body && !def.output) {
        throw new types_1.InternalServerError(`A response body was provided when none was expected`);
    }
    if (!output?.body && def.output) {
        throw new types_1.InternalServerError(`A response body is expected but none was provided`);
    }
    // mimetype
    if (def.output?.encoding &&
        (!output?.encoding ||
            !isValidEncoding(def.output?.encoding, output?.encoding))) {
        throw new types_1.InternalServerError(`Invalid response encoding: ${output?.encoding}`);
    }
    // output schema
    if (def.output?.schema) {
        try {
            const result = lexicons.assertValidXrpcOutput(nsid, output?.body);
            if (output) {
                output.body = result;
            }
        }
        catch (e) {
            throw new types_1.InternalServerError(e instanceof Error ? e.message : String(e));
        }
    }
}
function normalizeMime(v) {
    if (!v)
        return false;
    const fullType = mime_types_1.default.contentType(v);
    if (!fullType)
        return false;
    const shortType = fullType.split(';')[0];
    if (!shortType)
        return false;
    return shortType;
}
function isValidEncoding(possibleStr, value) {
    const possible = possibleStr.split(',').map((v) => v.trim());
    const normalized = normalizeMime(value);
    if (!normalized)
        return false;
    if (possible.includes('*/*'))
        return true;
    return possible.includes(normalized);
}
function getBodyPresence(req) {
    if (req.headers['transfer-encoding'] != null)
        return 'present';
    if (req.headers['content-length'] === '0')
        return 'empty';
    if (req.headers['content-length'] != null)
        return 'present';
    return 'missing';
}
function processBodyAsBytes(req) {
    return new Promise((resolve) => {
        const chunks = [];
        req.on('data', (chunk) => chunks.push(chunk));
        req.on('end', () => resolve(new Uint8Array(Buffer.concat(chunks))));
    });
}
function decodeBodyStream(req, maxSize) {
    const contentEncoding = req.headers['content-encoding'];
    const contentLength = req.headers['content-length'];
    const contentLengthParsed = contentLength
        ? parseInt(contentLength, 10)
        : undefined;
    if (Number.isNaN(contentLengthParsed)) {
        throw new types_1.XRPCError(xrpc_1.ResponseType.InvalidRequest, 'invalid content-length');
    }
    if (maxSize !== undefined &&
        contentLengthParsed !== undefined &&
        contentLengthParsed > maxSize) {
        throw new types_1.XRPCError(xrpc_1.ResponseType.PayloadTooLarge, 'request entity too large');
    }
    let transforms;
    try {
        transforms = (0, common_1.createDecoders)(contentEncoding);
    }
    catch (cause) {
        throw new types_1.XRPCError(xrpc_1.ResponseType.UnsupportedMediaType, 'unsupported content-encoding', undefined, { cause });
    }
    if (maxSize !== undefined) {
        const maxSizeChecker = new common_1.MaxSizeChecker(maxSize, () => new types_1.XRPCError(xrpc_1.ResponseType.PayloadTooLarge, 'request entity too large'));
        transforms.push(maxSizeChecker);
    }
    return transforms.length > 0
        ? (0, node_stream_1.pipeline)([req, ...transforms], () => { })
        : req;
}
function serverTimingHeader(timings) {
    return timings
        .map((timing) => {
        let header = timing.name;
        if (timing.duration)
            header += `;dur=${timing.duration}`;
        if (timing.description)
            header += `;desc="${timing.description}"`;
        return header;
    })
        .join(', ');
}
class ServerTimer {
    constructor(name, description) {
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: name
        });
        Object.defineProperty(this, "description", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: description
        });
        Object.defineProperty(this, "duration", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "startMs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
    start() {
        this.startMs = Date.now();
        return this;
    }
    stop() {
        (0, node_assert_1.default)(this.startMs, "timer hasn't been started");
        this.duration = Date.now() - this.startMs;
        return this;
    }
}
exports.ServerTimer = ServerTimer;
const parseReqNsid = (req) => (0, exports.parseUrlNsid)('originalUrl' in req ? req.originalUrl : req.url || '/');
exports.parseReqNsid = parseReqNsid;
/**
 * Validates and extracts the nsid from an xrpc path
 */
const parseUrlNsid = (url) => {
    // /!\ Hot path
    if (
    // Ordered by likelihood of failure
    url.length <= 6 ||
        url[5] !== '/' ||
        url[4] !== 'c' ||
        url[3] !== 'p' ||
        url[2] !== 'r' ||
        url[1] !== 'x' ||
        url[0] !== '/') {
        throw new types_1.InvalidRequestError('invalid xrpc path');
    }
    const startOfNsid = 6;
    let curr = startOfNsid;
    let char;
    let alphaNumRequired = true;
    for (; curr < url.length; curr++) {
        char = url.charCodeAt(curr);
        if ((char >= 48 && char <= 57) || // 0-9
            (char >= 65 && char <= 90) || // A-Z
            (char >= 97 && char <= 122) // a-z
        ) {
            alphaNumRequired = false;
        }
        else if (char === 45 /* "-" */ || char === 46 /* "." */) {
            if (alphaNumRequired) {
                throw new types_1.InvalidRequestError('invalid xrpc path');
            }
            alphaNumRequired = true;
        }
        else if (char === 47 /* "/" */) {
            // Allow trailing slash (next char is either EOS or "?")
            if (curr === url.length - 1 || url.charCodeAt(curr + 1) === 63) {
                break;
            }
            throw new types_1.InvalidRequestError('invalid xrpc path');
        }
        else if (char === 63 /* "?"" */) {
            break;
        }
        else {
            throw new types_1.InvalidRequestError('invalid xrpc path');
        }
    }
    // last char was one of: '-', '.', '/'
    if (alphaNumRequired) {
        throw new types_1.InvalidRequestError('invalid xrpc path');
    }
    // A domain name consists of minimum two characters
    if (curr - startOfNsid < 2) {
        throw new types_1.InvalidRequestError('invalid xrpc path');
    }
    // @TODO is there a max ?
    return url.slice(startOfNsid, curr);
};
exports.parseUrlNsid = parseUrlNsid;
