// @atproto/oauth-provider/dist/replay/replay-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplayManager = void 0;
const constants_js_1 = require("../constants.js");
const SECURITY_RATIO = 1.1; // 10% extra time for security
const asTimeFrame = (timeFrame) => Math.ceil(timeFrame * SECURITY_RATIO);
class ReplayManager {
    replayStore;
    constructor(replayStore) {
        this.replayStore = replayStore;
    }
    async uniqueAuth(jti, clientId) {
        return this.replayStore.unique(`Auth@${clientId}`, jti, asTimeFrame(constants_js_1.CLIENT_ASSERTION_MAX_AGE));
    }
    async uniqueJar(jti, clientId) {
        return this.replayStore.unique(`JAR@${clientId}`, jti, asTimeFrame(constants_js_1.JAR_MAX_AGE));
    }
    async uniqueDpop(jti, clientId) {
        return this.replayStore.unique(clientId ? `DPoP@${clientId}` : `DPoP`, jti, asTimeFrame(constants_js_1.DPOP_NONCE_MAX_AGE));
    }
    async uniqueCodeChallenge(challenge) {
        return this.replayStore.unique('CodeChallenge', challenge, asTimeFrame(constants_js_1.CODE_CHALLENGE_REPLAY_TIMEFRAME));
    }
}
exports.ReplayManager = ReplayManager;
