// @atproto/oauth-provider/dist/device/device-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceManager = exports.deviceManagerOptionsSchema = exports.keygripSchema = void 0;
const cookie_1 = require("cookie");
const zod_1 = require("zod");
const constants_js_1 = require("../constants.js");
const index_js_1 = require("../lib/http/index.js");
const request_js_1 = require("../lib/http/request.js");
const device_id_js_1 = require("./device-id.js");
const session_id_js_1 = require("./session-id.js");
/**
 * @see {@link https://www.npmjs.com/package/keygrip | Keygrip}
 */
exports.keygripSchema = zod_1.z.object({
    sign: zod_1.z.function().args(zod_1.z.any()).returns(zod_1.z.string()),
    verify: zod_1.z.function().args(zod_1.z.any(), zod_1.z.string()).returns(zod_1.z.boolean()),
    index: zod_1.z.function().args(zod_1.z.any(), zod_1.z.string()).returns(zod_1.z.number()),
});
exports.deviceManagerOptionsSchema = zod_1.z.object({
    /**
     * Controls whether the IP address is read from the `X-Forwarded-For` header
     * (if `true`), or from the `req.socket.remoteAddress` property (if `false`).
     */
    trustProxy: zod_1.z
        .function()
        .args(zod_1.z.string(), zod_1.z.number())
        .returns(zod_1.z.boolean())
        .optional(),
    /**
     * Amount of time (in ms) after which session IDs will be rotated
     *
     * @default 300e3 // (5 minutes)
     */
    rotationRate: zod_1.z.number().default(300e3),
    /**
     * Cookie options
     */
    cookie: zod_1.z
        .object({
        keys: exports.keygripSchema.optional(),
        /**
         * Name of the cookie used to identify the device
         *
         * @default 'session-id'
         */
        device: zod_1.z.string().default('device-id'),
        /**
         * Name of the cookie used to identify the session
         *
         * @default 'session-id'
         */
        session: zod_1.z.string().default('session-id'),
        /**
         * Url path for the cookie
         *
         * @default '/oauth/authorize'
         */
        path: zod_1.z.string().default('/oauth/authorize'),
        /**
         * Amount of time (in ms) after which the session cookie will expire.
         * If set to `null`, the cookie will be a session cookie (deleted when the
         * browser is closed).
         *
         * @default 10 years
         */
        age: zod_1.z
            .number()
            .nullable()
            .default(10 * 365.2 * 24 * 60 * 60e3),
        /**
         * Controls whether the cookie is only sent over HTTPS (if `true`), or also
         * over HTTP (if `false`). This should **NOT** be set to `false` in
         * production.
         */
        secure: zod_1.z.boolean().default(true),
        /**
         * Controls whether the cookie is sent along with cross-site requests.
         *
         * @default 'lax'
         */
        sameSite: zod_1.z.enum(['lax', 'strict']).default('lax'),
    })
        .default({}),
});
const cookieValueSchema = zod_1.z.tuple([device_id_js_1.deviceIdSchema, session_id_js_1.sessionIdSchema]);
/**
 * This class provides an abstraction for keeping track of DEVICE sessions. It
 * relies on a {@link DeviceStore} to persist session data and a cookie to
 * identify the session.
 */
class DeviceManager {
    store;
    options;
    constructor(store, options = {}) {
        this.store = store;
        this.options = exports.deviceManagerOptionsSchema.parse(options);
    }
    async load(req, res, forceRotate = false) {
        const cookie = await this.getCookie(req);
        if (cookie) {
            return this.refresh(req, res, cookie.value, forceRotate || cookie.mustRotate);
        }
        else {
            return this.create(req, res);
        }
    }
    async create(req, res) {
        const deviceMetadata = this.getRequestMetadata(req);
        const [deviceId, sessionId] = await Promise.all([
            (0, device_id_js_1.generateDeviceId)(),
            (0, session_id_js_1.generateSessionId)(),
        ]);
        await this.store.createDevice(deviceId, {
            sessionId,
            lastSeenAt: new Date(),
            userAgent: deviceMetadata.userAgent ?? null,
            ipAddress: deviceMetadata.ipAddress,
        });
        this.setCookie(res, [deviceId, sessionId]);
        return { deviceId, deviceMetadata };
    }
    async refresh(req, res, [deviceId, sessionId], forceRotate = false) {
        const data = await this.store.readDevice(deviceId);
        if (!data)
            return this.create(req, res);
        const lastSeenAt = new Date(data.lastSeenAt);
        const age = Date.now() - lastSeenAt.getTime();
        if (sessionId !== data.sessionId) {
            if (age <= constants_js_1.SESSION_FIXATION_MAX_AGE) {
                // The cookie was probably rotated by a concurrent request. Let's
                // update the cookie with the new sessionId.
                forceRotate = true;
            }
            else {
                // Something's wrong. Let's create a new session.
                await this.store.deleteDevice(deviceId);
                return this.create(req, res);
            }
        }
        const deviceMetadata = this.getRequestMetadata(req);
        if (forceRotate ||
            deviceMetadata.ipAddress !== data.ipAddress ||
            deviceMetadata.userAgent !== data.userAgent ||
            age > this.options.rotationRate) {
            await this.rotate(req, res, deviceId, {
                ipAddress: deviceMetadata.ipAddress,
                userAgent: deviceMetadata.userAgent || data.userAgent,
            });
        }
        return { deviceId, deviceMetadata };
    }
    async rotate(req, res, deviceId, data) {
        const sessionId = await (0, session_id_js_1.generateSessionId)();
        await this.store.updateDevice(deviceId, {
            ...data,
            sessionId,
            lastSeenAt: new Date(),
        });
        this.setCookie(res, [deviceId, sessionId]);
    }
    async getCookie(req) {
        const cookies = (0, index_js_1.parseHttpCookies)(req);
        if (!cookies)
            return null;
        const device = this.parseCookie(cookies, this.options.cookie.device, device_id_js_1.deviceIdSchema);
        const session = this.parseCookie(cookies, this.options.cookie.session, session_id_js_1.sessionIdSchema);
        // Silently ignore invalid cookies
        if (!device || !session) {
            // If the device cookie is valid, let's cleanup the DB
            if (device)
                await this.store.deleteDevice(device.value);
            return null;
        }
        return {
            value: [device.value, session.value],
            mustRotate: device.mustRotate || session.mustRotate,
        };
    }
    parseCookie(cookies, name, schema) {
        const result = schema.safeParse(cookies[name], { path: ['cookie', name] });
        if (!result.success)
            return null;
        const value = result.data;
        if (this.options.cookie.keys) {
            const hash = cookies[`${name}:hash`];
            if (!hash)
                return null;
            const idx = this.options.cookie.keys.index(value, hash);
            if (idx < 0)
                return null;
            return { value, mustRotate: idx !== 0 };
        }
        return { value, mustRotate: false };
    }
    setCookie(res, cookieValue) {
        this.writeCookie(res, this.options.cookie.device, cookieValue?.[0]);
        this.writeCookie(res, this.options.cookie.session, cookieValue?.[1]);
    }
    writeCookie(res, name, value) {
        const cookieOptions = {
            maxAge: value
                ? this.options.cookie.age == null
                    ? undefined
                    : this.options.cookie.age / 1000
                : 0,
            httpOnly: true,
            path: this.options.cookie.path,
            secure: this.options.cookie.secure !== false,
            sameSite: this.options.cookie.sameSite === 'lax' ? 'lax' : 'strict',
        };
        (0, index_js_1.appendHeader)(res, 'Set-Cookie', (0, cookie_1.serialize)(name, value || '', cookieOptions));
        if (this.options.cookie.keys) {
            (0, index_js_1.appendHeader)(res, 'Set-Cookie', (0, cookie_1.serialize)(`${name}:hash`, value ? this.options.cookie.keys.sign(value) : '', cookieOptions));
        }
    }
    getRequestMetadata(req) {
        return (0, request_js_1.extractRequestMetadata)(req, this.options);
    }
}
exports.DeviceManager = DeviceManager;
