// @atproto/aws/dist/kms.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KmsKeypair = void 0;
const aws = __importStar(require("@aws-sdk/client-kms"));
const secp256k1_1 = require("@noble/curves/secp256k1");
const key_encoder_1 = __importDefault(require("key-encoder"));
const ui8 = __importStar(require("uint8arrays"));
const crypto = __importStar(require("@atproto/crypto"));
const keyEncoder = new key_encoder_1.default('secp256k1');
class KmsKeypair {
    constructor(client, keyId, publicKey) {
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: client
        });
        Object.defineProperty(this, "keyId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: keyId
        });
        Object.defineProperty(this, "publicKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: publicKey
        });
        Object.defineProperty(this, "jwtAlg", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: crypto.SECP256K1_JWT_ALG
        });
    }
    static async load(cfg) {
        const { keyId, ...rest } = cfg;
        const client = new aws.KMS({
            ...rest,
            apiVersion: '2014-11-01',
        });
        const res = await client.getPublicKey({ KeyId: keyId });
        if (!res.PublicKey) {
            throw new Error('Could not find public key');
        }
        // public key comes back DER-encoded, so we translate it to raw 65 byte encoding
        const rawPublicKeyHex = keyEncoder.encodePublic(Buffer.from(res.PublicKey), 'der', 'raw');
        const publicKey = ui8.fromString(rawPublicKeyHex, 'hex');
        return new KmsKeypair(client, keyId, publicKey);
    }
    did() {
        return crypto.formatDidKey(this.jwtAlg, this.publicKey);
    }
    async sign(msg) {
        const res = await this.client.sign({
            KeyId: this.keyId,
            Message: msg,
            SigningAlgorithm: 'ECDSA_SHA_256',
        });
        if (!res.Signature) {
            throw new Error('Could not get signature');
        }
        // signature comes back DER encoded & not-normalized
        // we translate to raw 64 byte encoding
        // we also normalize s as no more than 1/2 prime order to pass strict verification
        // (prevents duplicating a signature)
        // more: https://github.com/bitcoin-core/secp256k1/blob/a1102b12196ea27f44d6201de4d25926a2ae9640/include/secp256k1.h#L530-L534
        const sig = secp256k1_1.secp256k1.Signature.fromDER(res.Signature);
        const normalized = sig.normalizeS();
        return normalized.toCompactRawBytes();
    }
}
exports.KmsKeypair = KmsKeypair;
exports.default = KmsKeypair;
