// @atproto/aws/dist/s3.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BlobStore = void 0;
const aws = __importStar(require("@aws-sdk/client-s3"));
const lib_storage_1 = require("@aws-sdk/lib-storage");
const crypto_1 = require("@atproto/crypto");
const repo_1 = require("@atproto/repo");
// @NOTE we use Upload rather than client.putObject because stream
// length is not known in advance. See also aws/aws-sdk-js-v3#2348.
class S3BlobStore {
    constructor(did, cfg) {
        Object.defineProperty(this, "did", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: did
        });
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "bucket", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "uploadTimeoutMs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        const { bucket, uploadTimeoutMs, ...rest } = cfg;
        this.bucket = bucket;
        this.uploadTimeoutMs = uploadTimeoutMs ?? 10000;
        this.client = new aws.S3({
            ...rest,
            apiVersion: '2006-03-01',
        });
    }
    static creator(cfg) {
        return (did) => {
            return new S3BlobStore(did, cfg);
        };
    }
    genKey() {
        return (0, crypto_1.randomStr)(32, 'base32');
    }
    getTmpPath(key) {
        return `tmp/${this.did}/${key}`;
    }
    getStoredPath(cid) {
        return `blocks/${this.did}/${cid.toString()}`;
    }
    getQuarantinedPath(cid) {
        return `quarantine/${this.did}/${cid.toString()}`;
    }
    async putTemp(bytes) {
        const key = this.genKey();
        // @NOTE abort results in error from aws-sdk "Upload aborted." with name "AbortError"
        const abortController = new AbortController();
        const timeout = setTimeout(() => abortController.abort(), this.uploadTimeoutMs);
        const upload = new lib_storage_1.Upload({
            client: this.client,
            params: {
                Bucket: this.bucket,
                Body: bytes,
                Key: this.getTmpPath(key),
            },
            // @ts-ignore native implementation fine in node >=15
            abortController,
        });
        try {
            await upload.done();
        }
        finally {
            clearTimeout(timeout);
        }
        return key;
    }
    async makePermanent(key, cid) {
        const alreadyHas = await this.hasStored(cid);
        if (!alreadyHas) {
            await this.move({
                from: this.getTmpPath(key),
                to: this.getStoredPath(cid),
            });
        }
        else {
            // already saved, so we no-op & just delete the temp
            await this.deleteKey(this.getTmpPath(key));
        }
    }
    async putPermanent(cid, bytes) {
        // @NOTE abort results in error from aws-sdk "Upload aborted." with name "AbortError"
        const abortController = new AbortController();
        const timeout = setTimeout(() => abortController.abort(), this.uploadTimeoutMs);
        const upload = new lib_storage_1.Upload({
            client: this.client,
            params: {
                Bucket: this.bucket,
                Body: bytes,
                Key: this.getStoredPath(cid),
            },
            // @ts-ignore native implementation fine in node >=15
            abortController,
        });
        try {
            await upload.done();
        }
        finally {
            clearTimeout(timeout);
        }
    }
    async quarantine(cid) {
        await this.move({
            from: this.getStoredPath(cid),
            to: this.getQuarantinedPath(cid),
        });
    }
    async unquarantine(cid) {
        await this.move({
            from: this.getQuarantinedPath(cid),
            to: this.getStoredPath(cid),
        });
    }
    async getObject(cid) {
        const res = await this.client.getObject({
            Bucket: this.bucket,
            Key: this.getStoredPath(cid),
        });
        if (res.Body) {
            return res.Body;
        }
        else {
            throw new repo_1.BlobNotFoundError();
        }
    }
    async getBytes(cid) {
        const res = await this.getObject(cid);
        return res.transformToByteArray();
    }
    async getStream(cid) {
        const res = await this.getObject(cid);
        return res;
    }
    async delete(cid) {
        await this.deleteKey(this.getStoredPath(cid));
    }
    async deleteMany(cids) {
        const keys = cids.map((cid) => this.getStoredPath(cid));
        await this.deleteManyKeys(keys);
    }
    async hasStored(cid) {
        return this.hasKey(this.getStoredPath(cid));
    }
    async hasTemp(key) {
        return this.hasKey(this.getTmpPath(key));
    }
    async hasKey(key) {
        try {
            const res = await this.client.headObject({
                Bucket: this.bucket,
                Key: key,
            });
            return res.$metadata.httpStatusCode === 200;
        }
        catch (err) {
            return false;
        }
    }
    async deleteKey(key) {
        await this.client.deleteObject({
            Bucket: this.bucket,
            Key: key,
        });
    }
    async deleteManyKeys(keys) {
        await this.client.deleteObjects({
            Bucket: this.bucket,
            Delete: {
                Objects: keys.map((k) => ({ Key: k })),
            },
        });
    }
    async move(keys) {
        try {
            await this.client.copyObject({
                Bucket: this.bucket,
                CopySource: `${this.bucket}/${keys.from}`,
                Key: keys.to,
            });
            await this.client.deleteObject({
                Bucket: this.bucket,
                Key: keys.from,
            });
        }
        catch (err) {
            handleErr(err);
        }
    }
}
exports.S3BlobStore = S3BlobStore;
const handleErr = (err) => {
    if (err?.['Code'] === 'NoSuchKey') {
        throw new repo_1.BlobNotFoundError();
    }
    else {
        throw err;
    }
};
exports.default = S3BlobStore;
