// @atproto/oauth-provider/dist/request/request-store-redis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestStoreRedis = void 0;
const redis_js_1 = require("../lib/redis.js");
const request_id_js_1 = require("./request-id.js");
class RequestStoreRedis {
    redis;
    constructor(options) {
        this.redis = (0, redis_js_1.createRedis)(options.redis);
    }
    async readRequest(id) {
        const data = await this.redis.get(id);
        return data ? JSON.parse(data) : null;
    }
    async createRequest(id, data) {
        const timeFrame = data.expiresAt.getTime() - Date.now();
        await this.redis.set(id, JSON.stringify(data), 'PX', timeFrame);
        if (data.code)
            await this.redis.set(data.code, id, 'PX', timeFrame);
    }
    async updateRequest(id, data) {
        const current = await this.readRequest(id);
        if (!current)
            throw new Error('Request not found');
        if (current.code)
            await this.redis.del(current.code);
        const newData = { ...current, ...data };
        await this.createRequest(id, newData);
    }
    async deleteRequest(id) {
        const data = await this.readRequest(id);
        if (!data)
            return;
        if (data.code)
            await this.redis.del(data.code);
        await this.redis.del(id);
    }
    async findRequestIdByCode(code) {
        const value = await this.redis.get(code);
        if (!value)
            return null;
        const parsed = request_id_js_1.requestIdSchema.safeParse(value);
        if (!parsed.success)
            return null;
        return parsed.data;
    }
    async findRequestByCode(code) {
        const id = await this.findRequestIdByCode(code);
        if (!id)
            return null;
        const data = await this.readRequest(id);
        if (!data)
            return null;
        return { id, data };
    }
}
exports.RequestStoreRedis = RequestStoreRedis;
